# Session 3

```{r, echo=FALSE, warning=FALSE}
library(knitr)
```

In this session we will learn:

* lists, another data structure
* Classes
* the DGEList object from the package limma

## Lists

Recall that we have previously learnt about the data structures: vectors, matrices and dataframes. Another important data structure is the list. Like a vector, it is 1 dimensional i.e. one row of data. Unlike vectors, you can put several data types in a list. Here, our list includes data of the integer, a character and a double types:

```{r}
list(1, "a", 1.5)
```


Not only can you put different data types into a list, you can also put a WHOLE data structure into one element of a list. In the list below, the first element is a vector that contains 3 numbers, the second element is a character and the third element is a dataframe that has two columns.


```{r}
list(c(1,2,3), 
     c("words", "letters"), 
     data.frame(column1 = c(1,2,3), column2 = c("a","b","c"))
     )
```

The output can often help you understand how the list is structured. The double brackets (e.g. `[[1]]`) signifies an element of the list and which index it is at. Here there are three elements in our list so the numbers in the double square brackets go from 1 to 3. Underneath the `[[1]]` and `[[2]]`, there is a `[1]` - this indicates that the first and second elements both contain a vector. Underneath `[[3]]` you see the standard output for a dataframe, which we have seen before.

You even include a list within a list (within a list, within a list....I call this 'list - ception'). This is where it starts to get confusing.

```{r}
list(1, list(1,"a"))
```

* The first element, indicated by `[[1]]`, is a vector, indicated by the `[1]` underneath.
* The second element, indicated by the first `[[2]]` contains a list:
    * `[[2]][[1]]` - tells you that the second element is a list, of which the first element of the inner list is the number 1.
    * `[[2]][[2]]` - tells you that the second element is a list, of which the second element of the inner list is "a".
    
Many bioconductor packages, including limma, use lists so it is an important data structure to understand.

**Challenge 3.1**

Below is the output from a list. 

1. How many element of the list are there?

2. Look carefully at each element and answer the following questions about EVERY element of the list:

  * How many elements does the element of the list contain? 
  * If there are several elements within this element, what does each element contain?


```{r, echo=FALSE}
list(
  list(c(1,2),  "b"), 
  c("a","b"),
  list(list("a", 1), "b"),
  "end"
  )
```

We can access elements of a list using square brackets. You may have noticed that there are often double square brackets `[[ ]]`. There is a subtle but important difference between single square brackets and double square brackets when subsetting lists.

Let's make a list:

```{r}
my_list <- list(
  c("a","b"),
  c(1,2,3),
  c(4,5,6)
  )
```

Take a look at the output of obtaining the first element with single square brackets and double square brackets:

```{r}
my_list[1]
```

```{r}
my_list[[1]]
```

The difference is minor; with `my_list[1]` there is an extra `[[1]]` at the top, with `my_list[[1]]` there is no `[[1]]` at the top.

The `str()` function gives us some more information:

```{r}
str(my_list[1])
```

```{r}
str(my_list[[1]])
```

This tells us that when you use single brackets, the output is a list containing one element, which is a vector. When you use double brackets, the output is just a vector, NOT contained within a list.

Let's use an analogy to gain a better understanding of this concept. Below is a picture of a cargo train, which contains a box in each segment. This represents a list containing 3 elements, with each element being the box.

```{r train, fig.cap="Cargo train representation of a list.", out.width="700px", fig.align='center', echo=FALSE}
knitr::include_graphics("Images/train.jpg")
```

Using a single bracket returns you the train segment with the box inside.

```{r trainS, fig.cap="Single brackets with our cargo train list.", out.width="300px", out.height="350px", fig.align='center', echo=FALSE}
knitr::include_graphics("Images/train_double.jpeg")
```

Using double brackets returns you just the box inside.

```{r trainD, fig.cap="Double brackets with our cargo train list.", out.width="300px", fig.align='center', echo=FALSE}
knitr::include_graphics("Images/box.png")
```

Back to our list:

```{r}
my_list
```

To obtain the first element of the vector contained within the first element of the list (the `"a"`) you can use:

```{r}
my_list[[1]][1]
```

The `[[1]]` gives you just the vector contained within the first element of the list:

```{r}
my_list[[1]]
```

The second square bracket `[1]` then gives you the first element of the vector:

```{r}
my_list[[1]][1]
```


**Challenge 3.2**

First create a new list:

```{r}
my_list2 <- list(
  c("a","b"),
  c(1,2,3),
  data.frame(Name = c("Sally", "Andy"), Age = c(40,32))
  )
my_list2
```

Subset this list to obtain a data structure that gives the following outputs:

1.
```{r, echo=FALSE}
my_list2[[1]]
```

2. 
```{r, echo=FALSE}
my_list2[[1]][2]
```

3.
```{r, echo=FALSE}
my_list2[2]
```

4. 
```{r, echo=FALSE}
my_list2[[3]][1,2]
```

5. 
```{r, echo=FALSE}
my_list2[[3]][1,]
```

6.
```{r, echo=FALSE}
my_list2[3]
```


### Named list

You can also have names for each element of your list:

```{r}
named_list <- list(
  name = c("Andy", "Sally"),
  age = c(32,40)
)
named_list
```

If your list is named, you can use `$` to access each element of your list:

```{r}
named_list$name
```

Note that the output is just the vector, NOT a vector within a list. Thus `named_list[[1]]` does the same thing as `named_list$name`.

Recall that we access columns in dataframes with the `$` symbol as well. This means that if you have a dataframe within a list, you can obtain a column within the dataframe that is within the list using two `$`'s: 

Let's start with a list that contains a dataframe as its first element
```{r}
named_list2 <- list(
 details = data.frame(name = c("Andy", "Sally"),
                      age = c(32,40))
)
named_list2
```

We can get the age column using:

```{r}
named_list2$details$age
```

*Extra*:The reason you can also access columns in a dataframe with `$` is because 'under the hood' a dataframe is actually a list. It is a list with the constraint that each element is a vector of the same length. Each element in the list is thus a column in the 'dataframe'.

**Challenge 3.3**

Take a look at the named list below:

```{r}
named_list3 <- list(
  cats = data.frame(name = c("Garfield", "Hello Kitty"),
                    age = c(3,10), stringsAsFactors = FALSE),
  dogs = data.frame(name = c("Spot", "Snoopy"),
                    age = c(5,14), stringsAsFactors = FALSE)
)
named_list3
```

Using just `$`, obtain:

1. the vector of cat names
2. the dog dataframe
3. the vector of dog ages


## Classes

Everything in R is an 'object' - every function that we have used and every data structure we have created in R. Each object falls under a 'class'. A class defines a type of object and what properties it has. For example, every list created with `list()` is an object that falls under the class 'list'. The 'list' class describes how 'list' objects behave. 


To find out what class an object falls under, use the function `class()`:

```{r}
class(list(1, 2, "a"))
class(list("a", "b", 4))
```

Bioconductor packages often use special classes. For example, the limma package uses the `DGEList` class. This is a class that is specifically designed for storing read count data from RNA sequencing. It is a special 'list' that must contain two components:

* `counts` - which must be a numeric matrix, that stores counts. Each row must be a gene and each column must be a sample.
* `samples` - which must be a dataframe, that contains information about each sample. Each row must be a sample and must contain information about the group (e.g. treatment group) the sample belongs to, the library size of that sample and the normalisation factor for that sample.

There are also a number of optional components of the `DGEList` class, such as a dataframe containing gene annotation information.

## Packages

Last session we installed the packages limma and edgeR. This downloads the files for each package and saves them to your computer. You generally only need to do this once.

To use a package you must 'load' them EACH time you start a new R session. You do this with the `library()` function. Let's load both limma and edgeR:

```{r, warning=FALSE}
library(edgeR)
library(limma)
```

## DGEList

The RNA sequencing analysis you will be guided through is a simplified version of that performed in the article from Law *et al.* [@law2016rna]. The RNA sequencing data we will use is from Sheridan *et al.* [@sheridan2015pooled]. It consists of 9 samples from 3 cell populations; basal, luminal progenitor (LP) and mature luminal (ML), which has been sorted from the mammary glands of female virgin mice. The reads have been aligned to the mouse reference genome (mm10) and reads summarised at the gene-level (using mm10 RefSeq-based annotation) to obtain gene counts. Gene level summarisation involves counting the number of reads mapped to each gene, for each sample. The resulting 'count of reads' is often referred to simply as 'counts'. 

We are going to start our RNA-seq analysis with gene counts. The data files for this session should have been emailed to you (though you can also obtain them from [Github](https://github.com/lucyleeow/BaseR_Intro)). Please uncompress (extract) the files and put them in your working directory.

Each data file corresponds to one sample and thus there is one data file for each sample. Each data file details the number of reads mapped to every gene for the sample corresponding to that data file. Within each data file, there are 3 columns - 'EntrezID', 'GeneLength' and 'Count.' 'EntrezID' and 'GeneLength' gives the EntrezID and gene length of one gene and 'Count' gives the number of reads mapped to that gene. The first four lines of one file (and thus one sample) is shown below:

```
EntrezID	GeneLength	Count
497097	3634	2
100503874	3259	0
100038431	1634	0
19888	9747	1
```

We will be looking at 9 samples (and using 9 data files) in total. Their details are shown below:


```{r samps, echo=FALSE}
files <- c("GSM1545535_10_6_5_11.txt", "GSM1545536_9_6_5_11.txt", 
   "GSM1545538_purep53.txt", "GSM1545539_JMS8-2.txt", 
   "GSM1545540_JMS8-3.txt", "GSM1545541_JMS8-4.txt", 
   "GSM1545542_JMS8-5.txt", "GSM1545544_JMS9-P7c.txt", 
   "GSM1545545_JMS9-P8c.txt")

samplenames <- c("10_6_5_11", "9_6_5_11", "purep53", "JMS8-2", "JMS8-3",
                 "JMS8-4", "JMS8-5", "JMS9-P7c", "JMS9-P8c")

kable(
  
  data.frame(`File name` = files,
   `Sample name` = samplenames,
   `Phenotype group` = c("LP", "ML", "Basal", "Basal", "ML", "LP", "Basal", "ML", "LP"),
   check.names = FALSE),
  caption = "Summary of samples."
  
)
```


To create a `DGEList` class object (or simply '`DGEList` object'), we will use the `readDGE()` function. There are three important arguments to this function:

* `files` - a vector of data file names
* `path` - the path to the directory that contains your data files. If the data files are in your working directory, don't worry about this argument. If the data files are somewhere else, like a folder called 'data', in your working directory you must give the path to that folder
* `colums` - the columns of the input files which have the gene names and counts respectively (as the column indices)

First, we will create a vector of the file names. You can simply copy and paste this code into your R script.

```{r}
files <- c("GSM1545535_10_6_5_11.txt", "GSM1545536_9_6_5_11.txt", 
   "GSM1545538_purep53.txt", "GSM1545539_JMS8-2.txt", 
   "GSM1545540_JMS8-3.txt", "GSM1545541_JMS8-4.txt", 
   "GSM1545542_JMS8-5.txt", "GSM1545544_JMS9-P7c.txt", 
   "GSM1545545_JMS9-P8c.txt")
```

Next, we will create our `DGEList` object. I have put my data files in a folder called "data" (within my working directory). Thus, I must specify `path = "data"`. Depending on where you have put your data files, you may need a different input to `path` or not have to include the `path` argument (if your data files are NOT within a folder in your working directory).

```{r}
x <- readDGE(files, path = "data/counts", columns = c(1,3))
```

The `readDGE()` function uses the read counts from our 9 data files (and thus 9 samples) to create a `DGEList` object containg count information for all 9 samples and every gene included in our data files.

It has 2 elements, one named `samples` and one named `counts`. You can take a look at each using `View()`:

```{r, eval=FALSE}
View(x$counts)
```

The output should look like this:

```{r dgecounts, fig.cap="View of the count matrix from the DGEList object we created and called x.", out.width="700px", fig.align='center', echo=FALSE}
knitr::include_graphics("Images/counts.png")
```

Notice that you can use the vertical and horizontal scroll bars to view all of the data. Each row is a gene, denoted by it's EntrezID and each column is a sample. Each value in this matrix gives the count value for one gene and one sample. Recall also that `DGEList` specifies that the `counts` element of the list must be a numeric matrix. 

Notice also, that the column names are the file names. The `DGEList` object was created using the information in the ".txt" files. The function `readDGE()` has used the file name as the name of each sample because we have not told `readDGE()` what the sample names are. It has therefore simply used the filenames. This makes logical sense as each file contained the count data for one sample.

The `samples` dataframe can also be viewed:

```{r, eval=FALSE}
View(x$samples)
```

The output should look something like this:

```{r dgesamples, fig.cap="View of the samples dataframe from the DGEList object we created and called x.", out.width="700px", fig.align='center', echo=FALSE}
knitr::include_graphics("Images/samplesdf.png")
```

This is a dataframe where each row is 1 sample, and details of each sample is given in the 4 columns. 

* `files` gives the file names.
* `group` details the phenotype group the sample belongs to. As we have not specified this yet, the default `1` is given for each sample.
* `lib.size` gives the library size. This is the total sum of all counts for that sample.
* `norm.factors` this gives the normalisation factor for each sample. As we have not calculated any normalisation factors, this is 1 for each sample.

This dataframe also has row names, which are currently the file names of each sample. Again, because we have not told `readDGE()` what the sample names are, it has used the file names.

Let's change the row names to be the sample names, instead of the full file name (see Table \@ref(tab:samps)).

We can do this by first creating a vector of sample names. You can simply copy and paste the code below into your R script.

```{r}
samplenames <- c("10_6_5_11", "9_6_5_11", "purep53", "JMS8-2", "JMS8-3",
                 "JMS8-4", "JMS8-5", "JMS9-P7c", "JMS9-P8c")
```

We can change the row names in the `samples` dataframe using the `rownames()` function. This function will give you the row names of a data structure (a dataframe in this instance):

```{r}
rownames(x$samples) 
```

We can see that the row names are the file names as we saw above. To replace these file names with the sample names we can run:

```{r}
rownames(x$samples) <- samplenames
```

We have seen this type of notation before in session 1 (section \@ref(vect)). This code assigns the `samplenames` vector as the row names of the `samples` dataframe.


**Challenge 3.4**

1. Use the `colnames()` function to replace the column names of the count matrix from being file names to sample names. 
We need to do this because the sample names in the `counts` matrix and the `samples` dataframe should be the same.
2. The code below creates a factor vector (called `group`) that specifies the phenotype group each sample belongs to. It is ordered such that the first element in the vector corresponds to the first row of the `samples` dataframe.

```{r}
group <- as.factor(c("LP", "ML", "Basal", "Basal", "ML", "LP", "Basal", "ML", "LP")) 
```

Replace the `group` column of the `samples` dataframe with the `group` factor vector.

*Hint*: you will need to use the list subsetting notation we learnt at the start of this session.

3. Additional gene annotation information about the genes from our RNA-seq data can be found in the file "Ses3_geneAnnot.tsv". Read this file in (specifying `stringsAsFactors = FALSE`) and add it as another element named `genes` in `x` (our `DGEList` object).

*Hint*: the command for adding an element to a list is similar to the command for adding a column to a dataframe. Take a look at Section \@ref(vect) to review how to do the latter.

## Homework

1. Subset the first column of the `count` matrix from `x` (our `DGEList` object) and 'save' (assign) it to a variable called `sample1`.
  + Find out how many genes had a count of '0'.
  + Find the total sum of all counts for that sample. Compare this number with the corresponding (first) number in the `lib.size` column from the `samples` dataframe. Is it the same?

```{r, eval=FALSE, echo=FALSE}
sample1 <- x$counts[,1]
str(sample1)
```

2. Calculate the total sum of the library sizes, using the `lib.size` column from the `samples` dataframe from `x` (our `DGEList` object).

```{r, eval=FALSE, echo=FALSE}
x$samples$lib.size
```

3. Using the `genes` dataframe from `x` (our `DGEList` object), find out how many genes are from chromosome 5.

```{r, eval=FALSE, echo=FALSE}
geneannot <- x$genes

geneannot <- geneannot[! is.na(geneannot$TXCHROM),]

chrm5 <- geneannot[geneannot$TXCHROM == "chr5",]
# 1324
```

*Hint*: you will need to remove rows where the `TXCHROM` column is `NA`. Revisit section \@ref(miss) to review missing values.

## Answers

**Challenge 3.1**

1. There are 4 elements of this list.
2. For each element -

* Within the first element, there are 2 elements. The first is a vector containing 2 numbers and the second is a vector containing one character type (note there are no 'scalars' in R, thus `"a"` is a vector with 1 element). 
* The second element contains 1 element. It is a vector containing two character types.
* The third element contains 2 elements. Within the first element is another list. Within this list there are 2 elements, both being character types. The second element of this nested list is a vector containing one character type.
* The fourth element is a vector containing one character type.


**Challenge 3.2**

1.
```{r}
my_list2[[1]]
```

2. 
```{r}
my_list2[[1]][2]
```

3.
```{r}
my_list2[2]
```

4. 
```{r}
my_list2[[3]][1,2]
```

5. 
```{r}
my_list2[[3]][1,]
```

6.
```{r}
my_list2[3]
```

**Challenge 3.3**

```{r}
# 1. the vector of cat names
named_list3$cats$name

# 2. the dog dataframe
named_list3$dogs

# 3. the vector of dog ages
named_list3$dogs$age
```

**Challenge 3.4**

```{r}
# 1
colnames(x$counts) <- samplenames
# 2
x$samples$group <- group
# 3
geneAnnot <- read.delim("data/Ses3_geneAnnot.tsv", stringsAsFactors = FALSE)
x$genes <- geneAnnot
```





