<!DOCTYPE html>
<html >

<head>

  <meta charset="UTF-8">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <title>Chapter 1 Session 1 | Introduction to Base R</title>
  <meta name="description" content="Introductory tutorial to base R, with a biological focus. In particular it will focus on analysing RNA-sequencing data with Bioconductor.">
  <meta name="generator" content="bookdown  and GitBook 2.6.7">

  <meta property="og:title" content="Chapter 1 Session 1 | Introduction to Base R" />
  <meta property="og:type" content="book" />
  
  
  <meta property="og:description" content="Introductory tutorial to base R, with a biological focus. In particular it will focus on analysing RNA-sequencing data with Bioconductor." />
  <meta name="github-repo" content="lucyleeow/BaseR_Intro" />

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Chapter 1 Session 1 | Introduction to Base R" />
  
  <meta name="twitter:description" content="Introductory tutorial to base R, with a biological focus. In particular it will focus on analysing RNA-sequencing data with Bioconductor." />
  

<meta name="author" content="Lucy Liu">


<meta name="date" content="2019-06-05">

  <meta name="viewport" content="width=device-width, initial-scale=1">
  <meta name="apple-mobile-web-app-capable" content="yes">
  <meta name="apple-mobile-web-app-status-bar-style" content="black">
  
  
<link rel="prev" href="index.html">
<link rel="next" href="session-2.html">
<script src="libs/jquery-2.2.3/jquery.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />







<script src="libs/kePrint-0.0.1/kePrint.js"></script>


<style type="text/css">
a.sourceLine { display: inline-block; line-height: 1.25; }
a.sourceLine { pointer-events: none; color: inherit; text-decoration: inherit; }
a.sourceLine:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode { white-space: pre; position: relative; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
code.sourceCode { white-space: pre-wrap; }
a.sourceLine { text-indent: -1em; padding-left: 1em; }
}
pre.numberSource a.sourceLine
  { position: relative; left: -4em; }
pre.numberSource a.sourceLine::before
  { content: attr(data-line-number);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; pointer-events: all; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {  }
@media screen {
a.sourceLine::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">Introduction to Base R</a></li>

<li class="divider"></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i>Preface</a><ul>
<li class="chapter" data-level="0.1" data-path="index.html"><a href="index.html#data-files"><i class="fa fa-check"></i><b>0.1</b> Data files</a></li>
</ul></li>
<li class="chapter" data-level="1" data-path="session-1.html"><a href="session-1.html"><i class="fa fa-check"></i><b>1</b> Session 1</a><ul>
<li class="chapter" data-level="1.1" data-path="session-1.html"><a href="session-1.html#r-and-rstudio"><i class="fa fa-check"></i><b>1.1</b> R and RStudio</a></li>
<li class="chapter" data-level="1.2" data-path="session-1.html"><a href="session-1.html#r-scripts"><i class="fa fa-check"></i><b>1.2</b> R scripts</a></li>
<li class="chapter" data-level="1.3" data-path="session-1.html"><a href="session-1.html#help"><i class="fa fa-check"></i><b>1.3</b> Help</a></li>
<li class="chapter" data-level="1.4" data-path="session-1.html"><a href="session-1.html#working-directory"><i class="fa fa-check"></i><b>1.4</b> Working directory</a></li>
<li class="chapter" data-level="1.5" data-path="session-1.html"><a href="session-1.html#maths"><i class="fa fa-check"></i><b>1.5</b> Maths</a></li>
<li class="chapter" data-level="1.6" data-path="session-1.html"><a href="session-1.html#comps"><i class="fa fa-check"></i><b>1.6</b> Comparisons</a></li>
<li class="chapter" data-level="1.7" data-path="session-1.html"><a href="session-1.html#variables"><i class="fa fa-check"></i><b>1.7</b> Variables</a><ul>
<li class="chapter" data-level="1.7.1" data-path="session-1.html"><a href="session-1.html#errors-and-warnings"><i class="fa fa-check"></i><b>1.7.1</b> Errors and warnings</a></li>
</ul></li>
<li class="chapter" data-level="1.8" data-path="session-1.html"><a href="session-1.html#data-structures"><i class="fa fa-check"></i><b>1.8</b> Data structures</a><ul>
<li class="chapter" data-level="1.8.1" data-path="session-1.html"><a href="session-1.html#vector"><i class="fa fa-check"></i><b>1.8.1</b> Vector</a></li>
<li class="chapter" data-level="1.8.2" data-path="session-1.html"><a href="session-1.html#matrix"><i class="fa fa-check"></i><b>1.8.2</b> Matrix</a></li>
<li class="chapter" data-level="1.8.3" data-path="session-1.html"><a href="session-1.html#dataframe"><i class="fa fa-check"></i><b>1.8.3</b> Dataframe</a></li>
</ul></li>
<li class="chapter" data-level="1.9" data-path="session-1.html"><a href="session-1.html#reading-in-data"><i class="fa fa-check"></i><b>1.9</b> Reading in data</a><ul>
<li class="chapter" data-level="1.9.1" data-path="session-1.html"><a href="session-1.html#summary"><i class="fa fa-check"></i><b>1.9.1</b> Summary</a></li>
<li class="chapter" data-level="1.9.2" data-path="session-1.html"><a href="session-1.html#structure"><i class="fa fa-check"></i><b>1.9.2</b> Structure</a></li>
</ul></li>
<li class="chapter" data-level="1.10" data-path="session-1.html"><a href="session-1.html#data-types"><i class="fa fa-check"></i><b>1.10</b> Data types</a><ul>
<li class="chapter" data-level="1.10.1" data-path="session-1.html"><a href="session-1.html#factors"><i class="fa fa-check"></i><b>1.10.1</b> Factors</a></li>
<li class="chapter" data-level="1.10.2" data-path="session-1.html"><a href="session-1.html#type-coercion"><i class="fa fa-check"></i><b>1.10.2</b> Type coercion</a></li>
</ul></li>
<li class="chapter" data-level="1.11" data-path="session-1.html"><a href="session-1.html#vect"><i class="fa fa-check"></i><b>1.11</b> Vectorisation</a></li>
<li class="chapter" data-level="1.12" data-path="session-1.html"><a href="session-1.html#write"><i class="fa fa-check"></i><b>1.12</b> Writing out data</a></li>
<li class="chapter" data-level="1.13" data-path="session-1.html"><a href="session-1.html#homework"><i class="fa fa-check"></i><b>1.13</b> Homework</a></li>
<li class="chapter" data-level="1.14" data-path="session-1.html"><a href="session-1.html#answers"><i class="fa fa-check"></i><b>1.14</b> Answers</a></li>
</ul></li>
<li class="chapter" data-level="2" data-path="session-2.html"><a href="session-2.html"><i class="fa fa-check"></i><b>2</b> Session 2</a><ul>
<li class="chapter" data-level="2.1" data-path="session-2.html"><a href="session-2.html#functions"><i class="fa fa-check"></i><b>2.1</b> Functions</a></li>
<li class="chapter" data-level="2.2" data-path="session-2.html"><a href="session-2.html#miss"><i class="fa fa-check"></i><b>2.2</b> Missing values</a><ul>
<li class="chapter" data-level="2.2.1" data-path="session-2.html"><a href="session-2.html#missing-codes"><i class="fa fa-check"></i><b>2.2.1</b> Missing codes</a></li>
</ul></li>
<li class="chapter" data-level="2.3" data-path="session-2.html"><a href="session-2.html#subsetting"><i class="fa fa-check"></i><b>2.3</b> Subsetting</a><ul>
<li class="chapter" data-level="2.3.1" data-path="session-2.html"><a href="session-2.html#subsetting-with-indicies"><i class="fa fa-check"></i><b>2.3.1</b> Subsetting with indicies</a></li>
<li class="chapter" data-level="2.3.2" data-path="session-2.html"><a href="session-2.html#subsetting-with-logicals"><i class="fa fa-check"></i><b>2.3.2</b> Subsetting with logicals</a></li>
<li class="chapter" data-level="2.3.3" data-path="session-2.html"><a href="session-2.html#in"><i class="fa fa-check"></i><b>2.3.3</b> %in%</a></li>
</ul></li>
<li class="chapter" data-level="2.4" data-path="session-2.html"><a href="session-2.html#merge"><i class="fa fa-check"></i><b>2.4</b> Merge</a></li>
<li class="chapter" data-level="2.5" data-path="session-2.html"><a href="session-2.html#homework-1"><i class="fa fa-check"></i><b>2.5</b> Homework</a></li>
<li class="chapter" data-level="2.6" data-path="session-2.html"><a href="session-2.html#answers-1"><i class="fa fa-check"></i><b>2.6</b> Answers</a></li>
</ul></li>
<li class="chapter" data-level="3" data-path="session-3.html"><a href="session-3.html"><i class="fa fa-check"></i><b>3</b> Session 3</a><ul>
<li class="chapter" data-level="3.1" data-path="session-3.html"><a href="session-3.html#lists"><i class="fa fa-check"></i><b>3.1</b> Lists</a><ul>
<li class="chapter" data-level="3.1.1" data-path="session-3.html"><a href="session-3.html#named-list"><i class="fa fa-check"></i><b>3.1.1</b> Named list</a></li>
</ul></li>
<li class="chapter" data-level="3.2" data-path="session-3.html"><a href="session-3.html#classes"><i class="fa fa-check"></i><b>3.2</b> Classes</a></li>
<li class="chapter" data-level="3.3" data-path="session-3.html"><a href="session-3.html#packages"><i class="fa fa-check"></i><b>3.3</b> Packages</a></li>
<li class="chapter" data-level="3.4" data-path="session-3.html"><a href="session-3.html#dgelist"><i class="fa fa-check"></i><b>3.4</b> DGEList</a></li>
<li class="chapter" data-level="3.5" data-path="session-3.html"><a href="session-3.html#homework-2"><i class="fa fa-check"></i><b>3.5</b> Homework</a></li>
<li class="chapter" data-level="3.6" data-path="session-3.html"><a href="session-3.html#answers-2"><i class="fa fa-check"></i><b>3.6</b> Answers</a></li>
</ul></li>
<li class="chapter" data-level="4" data-path="session-4.html"><a href="session-4.html"><i class="fa fa-check"></i><b>4</b> Session 4</a><ul>
<li class="chapter" data-level="4.1" data-path="session-4.html"><a href="session-4.html#filtering"><i class="fa fa-check"></i><b>4.1</b> Filtering</a></li>
<li class="chapter" data-level="4.2" data-path="session-4.html"><a href="session-4.html#normalisation"><i class="fa fa-check"></i><b>4.2</b> Normalisation</a></li>
<li class="chapter" data-level="4.3" data-path="session-4.html"><a href="session-4.html#mds"><i class="fa fa-check"></i><b>4.3</b> MDS</a><ul>
<li class="chapter" data-level="4.3.1" data-path="session-4.html"><a href="session-4.html#plotting"><i class="fa fa-check"></i><b>4.3.1</b> Plotting</a></li>
</ul></li>
<li class="chapter" data-level="4.4" data-path="session-4.html"><a href="session-4.html#saving-plots"><i class="fa fa-check"></i><b>4.4</b> Saving plots</a></li>
<li class="chapter" data-level="4.5" data-path="session-4.html"><a href="session-4.html#linear-modelling"><i class="fa fa-check"></i><b>4.5</b> Linear modelling</a><ul>
<li class="chapter" data-level="4.5.1" data-path="session-4.html"><a href="session-4.html#design-matrix"><i class="fa fa-check"></i><b>4.5.1</b> Design matrix</a></li>
<li class="chapter" data-level="4.5.2" data-path="session-4.html"><a href="session-4.html#contrasts"><i class="fa fa-check"></i><b>4.5.2</b> Contrasts</a></li>
<li class="chapter" data-level="4.5.3" data-path="session-4.html"><a href="session-4.html#linear-model-fitting"><i class="fa fa-check"></i><b>4.5.3</b> Linear model fitting</a></li>
</ul></li>
<li class="chapter" data-level="4.6" data-path="session-4.html"><a href="session-4.html#statistical-testing"><i class="fa fa-check"></i><b>4.6</b> Statistical testing</a></li>
<li class="chapter" data-level="4.7" data-path="session-4.html"><a href="session-4.html#answers-3"><i class="fa fa-check"></i><b>4.7</b> Answers</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="references.html"><a href="references.html"><i class="fa fa-check"></i>References</a></li>
<li class="divider"></li>
<li><a href="https://github.com/rstudio/bookdown" target="blank">Published with bookdown</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Introduction to Base R</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="session-1" class="section level1">
<h1><span class="header-section-number">Chapter 1</span> Session 1</h1>
<p>In this first session we will learn about:</p>
<ul>
<li>R and RStudio</li>
<li>data types and data structures</li>
<li>vectorisation</li>
<li>how to read in and write out data</li>
</ul>
<div id="r-and-rstudio" class="section level2">
<h2><span class="header-section-number">1.1</span> R and RStudio</h2>
<p>R is a free and open source statistical programming language, great for performing data analysis. RStudio is a free and open source R integrated development environment (IDE) which makes it easier for you to write code. It does this by providing you with auto-completion (of variable names, function names, file paths etc.), helping with formatting and keeping track of your variables.</p>
<p>You can think of R as the engine in a car and RStudio as the body &amp; controls. R is doing all the calculations/computations but RStudio makes it easier for you to use R.</p>
<p>When you first open RStudio, there will be three panels - see Figure <a href="session-1.html#fig:RS1">1.1</a> (don’t worry if your RStudio does not have the same colours/appearance - different versions of RStudio look slightly different and the text colour I use is not the default one.)</p>
<div class="figure"><span id="fig:RS1"></span>
<img src="Images/RS1.png" alt="Three panels in RStudio." width="700px" />
<p class="caption">
Figure 1.1: Three panels in RStudio.
</p>
</div>
<ul>
<li>Left panel: this panel features two tabs, ‘Console’, which is where you can type in commands for R to run and ‘Terminal’, which we won’t worry about in this course.</li>
<li>Top right panel:
<ul>
<li>Environment - in this tab you can see all variables you have created.</li>
<li>History - R keeps track of all commands you have run and you can review them in this tab.</li>
<li>Connections - this tab helps you connect to data sources but we will not be using it in this course.</li>
</ul></li>
<li>Bottom right:
<ul>
<li>Files - you can explore your file directory here and we will use it to set our working directory later.</li>
<li>Plots - plots that you create will either appear here or be saved to a file.</li>
<li>Help - help files for R functions can be viewed in this tab. Help files tell you about what a function does and how to use it.</li>
<li>Packages - basic R includes many useful functions. You can add even more functions by downloading packages. A package is a collection of functions, generally with a certain data analysis theme. For example, the package ‘limma’, which we will use later, includes functions for analysing RNA-sequencing data.</li>
<li>Viewer - this tab lets you view local web content but we won’t be using it in this course.</li>
</ul></li>
</ul>
</div>
<div id="r-scripts" class="section level2">
<h2><span class="header-section-number">1.2</span> R scripts</h2>
<p>To open a R script, go to File &gt; New File &gt; R Script.</p>
<div class="figure"><span id="fig:RS2"></span>
<img src="Images/RS2.png" alt="Opening a new R Script." width="700px" />
<p class="caption">
Figure 1.2: Opening a new R Script.
</p>
</div>
<p>This will open a fourth panel on the top left.</p>
<div class="figure"><span id="fig:RS3"></span>
<img src="Images/RS3.png" alt="Four panels, including a R Script." width="700px" />
<p class="caption">
Figure 1.3: Four panels, including a R Script.
</p>
</div>
<p>A R Script is just a text document. You can type and run commands using a R Script and down in the Console - the difference is that you can’t save the commands you run in the Console but you can save the R Script with all your commands. Note that to run a command in the Console press <code>Enter</code> key but to run a command in a R Script you must press <code>Cmd/Ctrl + Enter</code> keys.</p>
<p>Lastly, you can make ‘comments’ in your R Script. Comments are notes to yourself that are not interpreted by R and start with <code>#</code>:</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="co"># this is a note to myself</span>
<span class="dv">1</span><span class="op">+</span><span class="dv">3-2</span></code></pre>
<pre><code>## [1] 2</code></pre>
</div>
<div id="help" class="section level2">
<h2><span class="header-section-number">1.3</span> Help</h2>
<p>There are two ways to access help files in RStudio. You can type in the name of the function you want help with, in the top right of help tab (indicated with a red arrow):</p>
<div class="figure" style="text-align: center"><span id="fig:help"></span>
<img src="Images/help.png" alt="Accessing help files via 'help' tab." width="500px" />
<p class="caption">
Figure 1.4: Accessing help files via ‘help’ tab.
</p>
</div>
<p>The other option is to run the command <code>?</code> followed by the name of the function:</p>
<pre class="sourceCode r"><code class="sourceCode r">?sum</code></pre>
<p>Help files are very useful but can be difficult to interpret at first due to the technical language used. We won’t get too much practice reading help files during this course but I would encourage you to try to use them when figuring out how to use a new function.</p>
</div>
<div id="working-directory" class="section level2">
<h2><span class="header-section-number">1.4</span> Working directory</h2>
<p>Every file on your computer is located in a specific location. This location can be referred to by a path. In Mac, paths look something like this: <code>/Users/Lucy/Documents/</code>. In Windows, paths look something like this: <code>C:\Users\Lucy\Documents\</code>.</p>
<p>When you open a R session, it launches from a specific location. You can find out where this is using the command <code>getwd()</code>. This location called the ‘working directory’. R will, by default, look in this directory when reading in data and write out files/plots to this directory. It is often useful to have your data and R Scripts in the same directory and set this as your working directory.</p>
<p>You can set your working directory to be anywhere you like and we will now do this:</p>
<ol style="list-style-type: decimal">
<li>Make a folder for this course, somewhere sensible on your computer that you will be able to easily find.</li>
<li>Go back to your RStudio window, go to the bottom right panel, click on the ‘Files’ tab and then click on the three dots on the top right hand corner (Figure <a href="session-1.html#fig:wd1">1.5</a>).</li>
</ol>
<div class="figure" style="text-align: center"><span id="fig:wd1"></span>
<img src="Images/wd1.png" alt="Setting the working directory - 1." width="500px" />
<p class="caption">
Figure 1.5: Setting the working directory - 1.
</p>
</div>
<ol start="3" style="list-style-type: decimal">
<li>This will open up a new window (Figure <a href="session-1.html#fig:wd2">1.6</a>) which lets you explore the files and folders on your computer. Find the new folder you created, click on it then click ‘Open’.</li>
</ol>
<div class="figure" style="text-align: center"><span id="fig:wd2"></span>
<img src="Images/wd2.png" alt="Setting the working directory - 2." width="500px" />
<p class="caption">
Figure 1.6: Setting the working directory - 2.
</p>
</div>
<ol start="4" style="list-style-type: decimal">
<li>The files tab will now show the contents of your new folder (which should be empty). At the top of the files tab, click on More &gt; Set As Working Directory (Figure <a href="session-1.html#fig:wd3">1.7</a>).</li>
</ol>
<div class="figure" style="text-align: center"><span id="fig:wd3"></span>
<img src="Images/wd3.png" alt="Setting the working directory - 3." width="500px" />
<p class="caption">
Figure 1.7: Setting the working directory - 3.
</p>
</div>
<p>Please set your working directory to be this folder at the start of EVERY session.</p>
</div>
<div id="maths" class="section level2">
<h2><span class="header-section-number">1.5</span> Maths</h2>
<p>R performs maths and follows standard order of operations. In order of highest to lowest precedence, here is how mathematical operations are denoted in R:</p>
<ul>
<li><code>()</code> - parentheses</li>
<li><code>**</code> or <code>^</code> - exponents</li>
<li><code>/</code> - divide</li>
<li><code>*</code> - multiply</li>
<li><code>+</code> - add</li>
<li><code>-</code> - subtract</li>
</ul>
<p>Another useful function is modulus (<code>%%</code>), which gives the remainder after dividing:</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="dv">8</span><span class="op">%%</span><span class="dv">3</span></code></pre>
<pre><code>## [1] 2</code></pre>
</div>
<div id="comps" class="section level2">
<h2><span class="header-section-number">1.6</span> Comparisons</h2>
<p>You can also compare numbers in R:</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="dv">1</span> <span class="op">==</span><span class="st"> </span><span class="dv">1</span>  <span class="co"># equal to.</span></code></pre>
<pre><code>## [1] TRUE</code></pre>
<pre class="sourceCode r"><code class="sourceCode r"><span class="dv">1</span> <span class="op">!=</span><span class="st"> </span><span class="dv">1</span> <span class="co"># not equal to</span></code></pre>
<pre><code>## [1] FALSE</code></pre>
<pre class="sourceCode r"><code class="sourceCode r"><span class="dv">2</span> <span class="op">&gt;</span><span class="st"> </span><span class="dv">1</span>   <span class="co"># greater than.</span></code></pre>
<pre><code>## [1] TRUE</code></pre>
<pre class="sourceCode r"><code class="sourceCode r"><span class="dv">2</span> <span class="op">&lt;</span><span class="st"> </span><span class="dv">1</span>   <span class="co"># less than.</span></code></pre>
<pre><code>## [1] FALSE</code></pre>
<pre class="sourceCode r"><code class="sourceCode r"><span class="dv">1</span> <span class="op">&lt;=</span><span class="st"> </span><span class="dv">2</span>  <span class="co"># greater or equal to.</span></code></pre>
<pre><code>## [1] TRUE</code></pre>
<pre class="sourceCode r"><code class="sourceCode r"><span class="dv">1</span> <span class="op">&gt;=</span><span class="st"> </span><span class="dv">2</span>  <span class="co"># less than or equal to.</span></code></pre>
<pre><code>## [1] FALSE</code></pre>
<p>The <code>!</code> sign by itself means “not” and it reverses the logical. For example, “not” TRUE is FALSE.</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="op">!</span><span class="ot">TRUE</span></code></pre>
<pre><code>## [1] FALSE</code></pre>
<p>You can also compare words. What do you think will happen below?</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="st">&quot;cat&quot;</span> <span class="op">&gt;</span><span class="st"> &quot;dog&quot;</span></code></pre>
<pre><code>## [1] FALSE</code></pre>
<p>R will use alphabetical order to determine which word is ‘greater’:</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="st">&quot;a&quot;</span> <span class="op">&lt;</span><span class="st"> &quot;b&quot;</span></code></pre>
<pre><code>## [1] TRUE</code></pre>
</div>
<div id="variables" class="section level2">
<h2><span class="header-section-number">1.7</span> Variables</h2>
<p>A variable in R is just a name which refers to an ‘thing’ (more technically an ‘object’ in R).</p>
<p>For example, I can do some maths:</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="dv">2</span> <span class="op">+</span><span class="st"> </span><span class="dv">6</span> <span class="op">*</span><span class="st"> </span><span class="dv">7</span><span class="op">**</span><span class="dv">2</span></code></pre>
<pre><code>## [1] 296</code></pre>
<p>R outputs the result. However, if I want to ‘save’ this result to use later, I need to ‘assign’ the output to a variable. This can be thought of as giving it a name, so that we can refer to it later.</p>
<p>You can this with <code>&lt;-</code> (shortcut = <code>alt + -</code>) in R. (You can also use <code>=</code>, however stylistically <code>&lt;-</code> is preferred.)</p>
<p>Here R performs the calculation on the right of <code>&lt;-</code> and then saves the result as a variable called <code>my_num</code>.</p>
<pre class="sourceCode r"><code class="sourceCode r">my_num &lt;-<span class="st"> </span><span class="dv">2</span> <span class="op">+</span><span class="st"> </span><span class="dv">6</span> <span class="op">*</span><span class="st"> </span><span class="dv">7</span><span class="op">**</span><span class="dv">2</span></code></pre>
<p>Now if I run the command <code>my_num</code>, I see the number I stored earlier:</p>
<pre class="sourceCode r"><code class="sourceCode r">my_num</code></pre>
<pre><code>## [1] 296</code></pre>
<p>You can also ‘overwrite’ variables:</p>
<pre class="sourceCode r"><code class="sourceCode r">my_num &lt;-<span class="st"> </span><span class="dv">3</span> <span class="op">*</span><span class="st"> </span><span class="dv">4</span></code></pre>
<p>The above code evaluates <code>3 * 4</code> and assigns the output to the variable <code>my_num</code> - effectively ‘overwritting’ the previous value assigned to it.</p>
<pre class="sourceCode r"><code class="sourceCode r">my_num</code></pre>
<pre><code>## [1] 12</code></pre>
<p>R is particular about variable names. Variable names cannot:</p>
<ul>
<li>start with a number</li>
<li>contain any spaces</li>
</ul>
<p>If we try to create a variable that starts with a number, R will return an error:</p>
<pre class="sourceCode r"><code class="sourceCode r">2myvar &lt;-<span class="st"> </span><span class="dv">2</span> <span class="op">+</span><span class="st"> </span><span class="dv">6</span> <span class="op">*</span><span class="st"> </span><span class="dv">7</span><span class="op">**</span><span class="dv">2</span></code></pre>
<pre><code>## Error: &lt;text&gt;:1:2: unexpected symbol
## 1: 2myvar
##      ^</code></pre>
<div id="errors-and-warnings" class="section level3">
<h3><span class="header-section-number">1.7.1</span> Errors and warnings</h3>
<p>This is a good time to talk about ‘errors’ and ‘warnings’ in R.</p>
<p>An <em>error</em> is R telling you that it couldn’t do what you told it to do. Do not be disheartened at seeing an error message - it happens to everyone, including experienced programmers, all the time. DO read the error message, it is often very useful and tells you what you need to do to fix the error.</p>
<p>A <em>warning</em> is R telling you that is has done what you told it to do, however the result may not be what you want. Sometimes it is okay to ignore a warning, sometimes it is not!</p>
</div>
</div>
<div id="data-structures" class="section level2">
<h2><span class="header-section-number">1.8</span> Data structures</h2>
<p>A data structure can be thought of as a ‘container’ for data. There are a number of different data structures in R and each have different specifications about how data is stored.</p>
<p>Three commonly used data structures are summarised below:</p>
<div class="figure" style="text-align: center"><span id="fig:dstruct"></span>
<img src="Images/data-structure.png" alt="Data structures. Source: http://libguides.wellesley.edu." width="700px" />
<p class="caption">
Figure 1.8: Data structures. Source: <a href="http://libguides.wellesley.edu" class="uri">http://libguides.wellesley.edu</a>.
</p>
</div>
<p>Another useful data structure is a ‘list’, which we will talk about in Session 3.</p>
<div id="vector" class="section level3">
<h3><span class="header-section-number">1.8.1</span> Vector</h3>
<p>You can think of a vector like a row or column in excel. You can only store one <em>type</em> of data in a vector - e.g. all numbers or all text.</p>
<p>You can create vectors with the <code>c()</code> function (‘c’ for ‘combine’):</p>
<pre class="sourceCode r"><code class="sourceCode r">my_vect1 &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="dv">1</span>,<span class="dv">2</span>,<span class="dv">3</span>)
my_vect2 &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;a&quot;</span>, <span class="st">&quot;b&quot;</span>, <span class="st">&quot;c&quot;</span>)</code></pre>
</div>
<div id="matrix" class="section level3">
<h3><span class="header-section-number">1.8.2</span> Matrix</h3>
<p>A matrix is like an excel spreadsheet. It is two dimensional meaning you have columns and rows of data. You can only store one <em>type</em> of data in a matrix - e.g. all numbers or all text.</p>
</div>
<div id="dataframe" class="section level3">
<h3><span class="header-section-number">1.8.3</span> Dataframe</h3>
<p>Dataframes are also two dimensional (has both rows and columns) however, you can store different types of data in a dataframe. The only restriction is that all the data within one column must be of the same type.</p>
<p>A dataframe is composed of vectors, with each vector being one column. You can create a dataframe using the <code>data.frame()</code> function:</p>
<pre class="sourceCode r"><code class="sourceCode r">my_df &lt;-<span class="st"> </span><span class="kw">data.frame</span>(
  <span class="dt">cats =</span> <span class="kw">c</span>(<span class="st">&quot;Hello Kitty&quot;</span>, <span class="st">&quot;Garfield&quot;</span>), 
  <span class="dt">weight =</span> <span class="kw">c</span>(<span class="fl">4.5</span>,<span class="dv">7</span>)
)
my_df</code></pre>
<pre><code>##          cats weight
## 1 Hello Kitty    4.5
## 2    Garfield    7.0</code></pre>
<p>There are a few important things to note:</p>
<ol style="list-style-type: decimal">
<li><code>name</code> and <code>age</code> become the names of the columns.</li>
<li>the <code>,</code> at the end of <code>cats = c(&quot;Hello Kitty&quot;, &quot;Garfield&quot;)</code> is important and should not be missed.</li>
<li>each column in a dataframe is essentially a vector. Do not forget the <code>c()</code> when inputing the values within each column.</li>
</ol>
<p>You can access a column in a dataframe with the shortcut <code>$</code>. Notice that the names of all columns of the dataframe appears after typing in <code>my_df$</code>:</p>
<div class="figure" style="text-align: center"><span id="fig:dfdollar"></span>
<img src="Images/df_dollar.png" alt="Accessing columns in a dataframe using the '$' symbol." width="350px" />
<p class="caption">
Figure 1.9: Accessing columns in a dataframe using the ‘$’ symbol.
</p>
</div>
<p><strong>Challenge 1.1</strong></p>
<p>Create a dataframe called <code>my_df2</code> with 2 columns. The first column named “dogs” should be a vector with the values: “spot”, “snoopy”. The second column named “weight” should be a vector with the following values: 3.5, 4.8.</p>
<p>Write this dataframe out to a .tsv file named ‘Ses1_dogs.tsv’.</p>
</div>
</div>
<div id="reading-in-data" class="section level2">
<h2><span class="header-section-number">1.9</span> Reading in data</h2>
<p>Let’s read in some data and start exploring it.</p>
<p>You should have received the data files via email before the course. Please download these files and make sure they are located in your working directory. Recall, we set our working directory above. You can find the location of your working directory with the function <code>getwd()</code>.</p>
<p>If you are following along outside of the WEHI course, you can download the data files from <a href="https://github.com/lucyleeow/BaseR_Intro/tree/master/data">Github</a> - instructions for downloading data from GitHub can be found in the Preface.</p>
<p>The file we want to read in is named ‘Ses1_genes.tsv’.</p>
<p>I have put all my data files in a directory called ‘data’ - thus the path to the file (relative to my working directory) is ‘data/Ses1_genes.tsv’. Depending on where you have put your data (in your working directory or in another file in your working directory), the path to your file ‘Ses1_genes.tsv’ may be different.</p>
<p>Read in your data by typing in the path to your ‘Ses1_genes.tsv’ file (relative to your working directory), within the brackets <code>( )</code>:</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">read.delim</span>(<span class="st">&quot;data/Ses1_genes.tsv&quot;</span>)</code></pre>
<pre><code>##     SYMBOL GeneLength Count
## 1  Gm10568       1634     0
## 2  Gm19860        799     4
## 3  Gm19938       3259     0
## 4   Lypla1       2433   768
## 5      Rp1       9747     0
## 6    Sox17       3130     1
## 7    Tcea1       2847   810
## 8   Mrpl15       4203   431
## 9     Xkr4       3634     1
## 10   Rgs20       2241   452</code></pre>
<p>Note that <code>read.delim()</code> also lets you specify what kind of file you are reading in - by this I mean how each value is separated. Two common formats are ‘csv’ (comma separated values) and ‘tsv’ (tab separated values).</p>
<p>Example of csv file:</p>
<pre><code>Name, Age
Andy, 10
Bob, 8</code></pre>
<p>Example of tsv file:</p>
<pre><code>Name  Age
Andy  10
Bob   8</code></pre>
<p>Notice in the above two examples the values are separated by different ‘characters’.</p>
<p>You can specify what ‘character’ separates each value by using the <code>sep</code> input in <code>read.delim()</code>. E.g. if your file was a csv, you can read it in using:</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">read.delim</span>(<span class="st">&quot;file.csv&quot;</span>, <span class="dt">sep =</span> <span class="st">&quot;,&quot;</span>)</code></pre>
<p>We don’t need to specify the <code>sep</code> for our file, which is a tsv, as the default separater <code>read.delim()</code> is tab. We only need to specify the separater character when we are reading in a file NOT separated by tabs.</p>
<p>Above, we have read in our data as a dataframe and printed it. However, we can’t refer to this dataframe again and manipulate it because we haven’t assigned it to a variable.</p>
<p>Let’s assign our dataframe to a variable called <code>genes</code>:</p>
<pre class="sourceCode r"><code class="sourceCode r">genes &lt;-<span class="st"> </span><span class="kw">read.delim</span>(<span class="st">&quot;data/Ses1_genes.tsv&quot;</span>)</code></pre>
<p>Notice how <code>genes</code> now appears in our ‘Environment’ tab:</p>
<div class="figure" style="text-align: center"><span id="fig:import"></span>
<img src="Images/import.png" alt="Variable in 'Environment' tab." width="500px" />
<p class="caption">
Figure 1.10: Variable in ‘Environment’ tab.
</p>
</div>
<p>All variables that we create will be shown in this tab, so it is a useful way to keep track of variables that we have created. Notice how R also tells us that there are 10 observations (rows) and 3 variables (columns) in the <code>genes</code> dataframe.</p>
<div id="summary" class="section level3">
<h3><span class="header-section-number">1.9.1</span> Summary</h3>
<p>A useful function for investigating your data is <code>summary()</code>. Running this function on our <code>genes</code> dataframe provides us with summary statistics on all the numeric columns. For the columns that don’t contain numbers, the output simply states the length of that column.</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">summary</span>(genes)</code></pre>
<pre><code>##      SYMBOL    GeneLength       Count       
##  Gm10568:1   Min.   : 799   Min.   :  0.00  
##  Gm19860:1   1st Qu.:2289   1st Qu.:  0.25  
##  Gm19938:1   Median :2988   Median :  2.50  
##  Lypla1 :1   Mean   :3393   Mean   :246.70  
##  Mrpl15 :1   3rd Qu.:3540   3rd Qu.:446.75  
##  Rgs20  :1   Max.   :9747   Max.   :810.00  
##  (Other):4</code></pre>
</div>
<div id="structure" class="section level3">
<h3><span class="header-section-number">1.9.2</span> Structure</h3>
<p>A useful function for understanding how our data is stored in R is <code>str()</code> (structure).</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">str</span>(genes)</code></pre>
<pre><code>## &#39;data.frame&#39;:    10 obs. of  3 variables:
##  $ SYMBOL    : Factor w/ 10 levels &quot;Gm10568&quot;,&quot;Gm19860&quot;,..: 1 2 3 4 7 8 9 5 10 6
##  $ GeneLength: int  1634 799 3259 2433 9747 3130 2847 4203 3634 2241
##  $ Count     : int  0 4 0 768 0 1 810 431 1 452</code></pre>
<p>The output tells us that <code>genes</code> is a dataframe. It also tells us what <em>data type</em> each column is.</p>
</div>
</div>
<div id="data-types" class="section level2">
<h2><span class="header-section-number">1.10</span> Data types</h2>
<p>Every bit of data in R has a ‘data type label’. The label doesn’t change the data in any way - it just tells R what kind of data it is and thus what it can and can’t do with the data.</p>
<p>For example, it makes sense to perform mathematical functions on numbers but not on words. It makes sense to look for a certain term, like a gene name, in words but not in numbers.</p>
<p>For example, you can’t add words so the code belwo doesn’t work. Note that quotes (<code>&quot; &quot;</code> or <code>' '</code>) ALWAYS surround values that are of the ‘character’ data type:</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="st">&quot;one&quot;</span> <span class="op">+</span><span class="st"> &quot;two&quot;</span></code></pre>
<pre><code>## Error in &quot;one&quot; + &quot;two&quot;: non-numeric argument to binary operator</code></pre>
<p>This does work because they are numbers:</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="dv">1</span> <span class="op">+</span><span class="st"> </span><span class="dv">2</span></code></pre>
<pre><code>## [1] 3</code></pre>
<p>There are five basic types of data in R:</p>
<ul>
<li><code>logical</code> - either TRUE or FALSE. This is useful for data that only has two possible values, like if patient has a disease or not.</li>
<li><code>integer</code> - number without decimal point e.g. <code>3</code>.</li>
<li><code>double</code> - number with decimal point e.g. <code>3.14</code>.</li>
<li><code>complex</code> - complex number with a real and imaginary part e.g. <code>1 + 3i</code></li>
<li><code>character</code> - Anything with character(s) within it. Quotes (double or single) signify this data type e.g. <code>&quot;pi&quot;</code>. Also known as a ‘string’.</li>
</ul>
<p>In our <code>genes</code> dataframe above, we can see that both the <code>GeneLength</code> and <code>Count</code> columns are integers (‘int’). But what is the <code>SYMBOL</code> column? It is a ‘Factor’.</p>
<div id="factors" class="section level3">
<h3><span class="header-section-number">1.10.1</span> Factors</h3>
<p>Factor is a data type label used for categorical variables (e.g. small, medium and large OR red, blue and green). Internally, R stores factors as numbers (integers), with each number corresponding to a category.</p>
<p>For example, if your data was:</p>
<pre><code>red, blue, green, red, green</code></pre>
<p>Internally, R would store the data as:</p>
<pre><code>1, 2, 3, 1, 3</code></pre>
<p>Each number corresponds to a category. This information is also stored. In R, the categorical values each number corresponds to, is called ‘levels’.</p>
<p>The levels for the above data would be:</p>
<pre><code>1 = red
2 = blue
3 = green</code></pre>
<p>Factors can be difficult to work with, so we don’t want our gene symbols to be ‘labelled’ as factors. We can do this by specifying this when reading our data in:</p>
<pre class="sourceCode r"><code class="sourceCode r">genes &lt;-<span class="st"> </span><span class="kw">read.delim</span>(<span class="st">&quot;data/Ses1_genes.tsv&quot;</span>, <span class="dt">stringsAsFactors =</span> <span class="ot">FALSE</span>)</code></pre>
<p>By default, R will label all word (‘character’) data as a ‘factor’. Setting <code>stringsAsFactors</code> to be <code>FALSE</code>, tells R that you DON’T want it to do this.</p>
<p>We can check the structure again:</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">str</span>(genes)</code></pre>
<pre><code>## &#39;data.frame&#39;:    10 obs. of  3 variables:
##  $ SYMBOL    : chr  &quot;Gm10568&quot; &quot;Gm19860&quot; &quot;Gm19938&quot; &quot;Lypla1&quot; ...
##  $ GeneLength: int  1634 799 3259 2433 9747 3130 2847 4203 3634 2241
##  $ Count     : int  0 4 0 768 0 1 810 431 1 452</code></pre>
<p>Notice that now the column <code>SYMBOL</code> is now a character (‘chr’).</p>
<p>Another way to change the labels of data in R is with the following functions:</p>
<ul>
<li><code>as.logical()</code></li>
<li><code>as.integer()</code></li>
<li><code>as.double()</code></li>
<li><code>as.complex()</code></li>
<li><code>as.character()</code></li>
</ul>
<p>For example, this will turn a integer vector into a character vector:</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">as.character</span>(<span class="kw">c</span>(<span class="dv">1</span>,<span class="dv">2</span>,<span class="dv">3</span>))</code></pre>
<pre><code>## [1] &quot;1&quot; &quot;2&quot; &quot;3&quot;</code></pre>
</div>
<div id="type-coercion" class="section level3">
<h3><span class="header-section-number">1.10.2</span> Type coercion</h3>
<p>In some circumstances, R will change the data type label of your data. This is called ‘type coercion’. A common scenario in which this will happen is with vector (and dataframe column) labeling.</p>
<p>All elements within a vector (and within a column of a dataframe) must have the same data type label. Thus, if you create a vector like the one below, the whole vector will be labelled as one data type. Which data type do you think it will be?</p>
<pre class="sourceCode r"><code class="sourceCode r">my_vect3 &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="dv">1</span>, <span class="dv">2</span>, <span class="st">&quot;a&quot;</span>)</code></pre>
<p>We can find out with <code>str()</code>:</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">str</span>(my_vect3)</code></pre>
<pre><code>##  chr [1:3] &quot;1&quot; &quot;2&quot; &quot;a&quot;</code></pre>
<p>It has labelled the whole vector as characters. This is because the letter <code>&quot;a&quot;</code> cannot be “expressed” as a number but numbers can be “expressed” as characters - thus R will always pick the data type label that does not result in any loss of information. Recall that data type labels NEVER change the data - only the label changes.</p>
<p>There is an order to which data label R will label a vector as. Of all the data types that exist in the vector, the data type of the element that is furthermost right, will be the “final” data type label for the vector.</p>
<p><code>logical</code> &gt; <code>integer</code> &gt; <code>double</code> &gt; <code>complex</code> &gt; <code>character</code></p>
<p>This order makes sense:</p>
<ul>
<li>the logicals <code>TRUE</code> and <code>FALSE</code> can be represented by numbers with <code>TRUE</code> = <code>1</code> and <code>FALSE</code> = <code>0</code>.
<ul>
<li>when logicals are labelled as the character type, <code>TRUE</code> simply becomes <code>&quot;TRUE&quot;</code></li>
</ul></li>
<li>an integer can easily be represented by a double - <code>3</code> becomes <code>3.0</code></li>
<li>as we saw above, numbers can easily be represented as a character</li>
</ul>
<p>The last thing to note is that you may see the type <code>num</code>:</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">str</span>(<span class="kw">c</span>(<span class="dv">1</span>,<span class="dv">2</span>,<span class="dv">3</span>))</code></pre>
<pre><code>##  num [1:3] 1 2 3</code></pre>
<p><code>num</code> stands for numeric and is just the ‘number’ data types grouped together. R does this for two reasons. First, generally you don’t need to know exactly what ‘type’ of number class your data is. Second, R will often convert between number types depending the type of calculations/function performed.</p>
<p><strong>Challenge 1.3</strong></p>
<ol style="list-style-type: decimal">
<li>What data type will the following vectors be?</li>
</ol>
<pre class="sourceCode r"><code class="sourceCode r"><span class="co"># 1</span>
<span class="kw">c</span>(<span class="ot">TRUE</span>, <span class="dv">3</span>, <span class="dv">4</span>)
<span class="co"># 2 </span>
<span class="kw">c</span>(<span class="st">&quot;hello&quot;</span>, <span class="ot">TRUE</span>, <span class="dv">4</span>)
<span class="co"># 3</span>
<span class="kw">c</span>(<span class="dv">4</span>, <span class="st">&quot;5&quot;</span>)</code></pre>
<ol start="2" style="list-style-type: decimal">
<li>Create a vector called <code>my_vect3</code> that contains the following numbers: 1,1,2,3,5,8.
<ul>
<li>Convert this vector into the character data type. How can you tell that the values are now characters?</li>
</ul></li>
</ol>
</div>
</div>
<div id="vect" class="section level2">
<h2><span class="header-section-number">1.11</span> Vectorisation</h2>
<p>R is quite efficient at doing calculations or performing a functions on a vector of data.</p>
<p>Let’s use the <code>Count</code> column from our dataframe <code>genes</code>. We can access just one column using the <code>$</code> symbol:</p>
<pre class="sourceCode r"><code class="sourceCode r">genes<span class="op">$</span>Count</code></pre>
<pre><code>##  [1]   0   4   0 768   0   1 810 431   1 452</code></pre>
<p>What do you think will happen if we do this?</p>
<pre class="sourceCode r"><code class="sourceCode r">genes<span class="op">$</span>Count <span class="op">+</span><span class="st"> </span><span class="dv">10</span></code></pre>
<pre><code>##  [1]  10  14  10 778  10  11 820 441  11 462</code></pre>
<p>R will perform this calculation on all numbers in the <code>Count</code> column and return a vector.</p>
<p>We can add this new vector to our dataframe, as a new column.</p>
<pre class="sourceCode r"><code class="sourceCode r">genes<span class="op">$</span>Count_<span class="dv">2</span> &lt;-<span class="st"> </span>genes<span class="op">$</span>Count <span class="op">+</span><span class="st"> </span><span class="dv">10</span></code></pre>
<p>There are a few things happening in the above code:</p>
<ol style="list-style-type: decimal">
<li>The code on the right side of <code>&lt;-</code> is evaluated first. It returns a vector as we saw above.</li>
<li>This vector is assigned to a column in <code>genes</code> called <code>Count_2</code>. Since this column does not yet exist in the <code>genes</code> dataframe, a new column called <code>Count_2</code> is created first. (If there was already a column name <code>Count_2</code>, this command will have overwritten that column with the new vector created in the right side of <code>&lt;-</code>)</li>
</ol>
<p>Note that you could also REPLACE a column using the same notation. For example, if, in the command above, the left side was <code>genes$Count</code>, the old column called <code>Count</code> would be REPLACED with the new vector of numbers created on the right side.</p>
<p><strong>Challenge 1.2</strong></p>
<p>Create a new column called <code>Prop_Count</code> that contains each count value as a proportion of the total count value of all 10 genes in the dataframe. E.g. if <code>Count</code> was 10 and total count of all 10 genes is 100, that row in <code>Prop_Count</code> should be 0.1.</p>
<p><em>Hint</em> use the <code>sum()</code> function.</p>
</div>
<div id="write" class="section level2">
<h2><span class="header-section-number">1.12</span> Writing out data</h2>
<p>The last thing we will do this session is learn to write out data using the function <code>write.table()</code>.</p>
<p>There are a few things we must tell <code>write.table()</code>, for it to be able to write out the data the way we want:</p>
<ul>
<li><code>x</code> - the name of this input is not very informative, but first you must tell the function what you want to write out. In our case we want to write out our dataframe <code>genes</code>.</li>
<li><code>file</code> - the name of the file that we want to write to.</li>
<li><code>sep</code> - how each value in our output file is separated. Common file formats are ‘csv’ and ‘tsv’ (discussed above). In R, a tab is represented by <code>&quot;\t&quot;</code>.</li>
<li><code>row.names</code> - this is either <code>TRUE</code> or <code>FALSE</code>, and let’s you specify whether you want to write out row names. If your dataframe does not have row names, put <code>FALSE</code>.</li>
<li><code>col.names</code> - this is also either <code>TRUE</code> or <code>FALSE</code>, and let’s you specify whether you want to write out column names. If your dataframe has column names, put <code>TRUE</code>.</li>
</ul>
<p>We can write out our <code>genes</code> dataframe into a .tsv file using the command below:</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">write.table</span>(<span class="dt">x =</span> genes, <span class="dt">file =</span> <span class="st">&quot;Ses1_Genes_output.tsv&quot;</span>, <span class="dt">sep =</span> <span class="st">&quot;</span><span class="ch">\t</span><span class="st">&quot;</span>,
            <span class="dt">row.names =</span> <span class="ot">FALSE</span>, <span class="dt">col.names =</span> <span class="ot">TRUE</span>)</code></pre>
</div>
<div id="homework" class="section level2">
<h2><span class="header-section-number">1.13</span> Homework</h2>
<p>Read in the data file “Ses1_homework.tsv” using <code>read.delim()</code>. This file is similar to the “Ses1_genes.tsv” file but with a different 10 genes.</p>
<p>Create a new column called <code>Prop_GeneLength</code> that contains the <code>Count</code> value as a proportion of <code>GeneLength</code>. E.g. if the count value was 50 and the gene length was 2400, the <code>Prop_GeneLength</code> value for that row would be 0.02083333.</p>
<p>Finally, write out this new dataframe as a file called “Ses1_homework_output.tsv”, as a tsv (tab separated values) file.</p>
</div>
<div id="answers" class="section level2">
<h2><span class="header-section-number">1.14</span> Answers</h2>
<p><strong>Challenge 1.1</strong></p>
<pre class="sourceCode r"><code class="sourceCode r">my_df2 &lt;-<span class="st"> </span><span class="kw">data.frame</span>(
  <span class="dt">dogs =</span> <span class="kw">c</span>(<span class="st">&quot;Garfield&quot;</span>, <span class="st">&quot;Hello Kitty&quot;</span>),
  <span class="dt">weight =</span> <span class="kw">c</span>(<span class="fl">3.5</span>, <span class="fl">4.8</span>)
)</code></pre>
<p><strong>Challenge 1.2</strong></p>
<ol style="list-style-type: decimal">
<li></li>
</ol>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">str</span>(<span class="kw">c</span>(<span class="ot">TRUE</span>, <span class="dv">3</span>, <span class="dv">4</span>))</code></pre>
<pre><code>##  num [1:3] 1 3 4</code></pre>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">str</span>(<span class="kw">c</span>(<span class="st">&quot;hello&quot;</span>, <span class="ot">TRUE</span>, <span class="dv">4</span>))</code></pre>
<pre><code>##  chr [1:3] &quot;hello&quot; &quot;TRUE&quot; &quot;4&quot;</code></pre>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">str</span>(<span class="kw">c</span>(<span class="dv">4</span>, <span class="st">&quot;5&quot;</span>))</code></pre>
<pre><code>##  chr [1:2] &quot;4&quot; &quot;5&quot;</code></pre>
<ol start="2" style="list-style-type: decimal">
<li></li>
</ol>
<pre class="sourceCode r"><code class="sourceCode r">my_vect3 &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="dv">1</span>,<span class="dv">1</span>,<span class="dv">2</span>,<span class="dv">3</span>,<span class="dv">5</span>,<span class="dv">8</span>)
<span class="co"># convert to character</span>
my_vect3 &lt;-<span class="st"> </span><span class="kw">as.character</span>(my_vect3)
<span class="co"># quotes around each value show that they are of the character type</span>
my_vect3 </code></pre>
<pre><code>## [1] &quot;1&quot; &quot;1&quot; &quot;2&quot; &quot;3&quot; &quot;5&quot; &quot;8&quot;</code></pre>
<p><strong>Challenge 1.3</strong></p>
<p>The <code>sum()</code> function can be used to calculate the total Count of all 10 rows by giving the <code>genes$Count</code> vector as the input to <code>sum()</code>.</p>
<pre class="sourceCode r"><code class="sourceCode r">genes<span class="op">$</span>Prop_Count &lt;-<span class="st"> </span>genes<span class="op">$</span>Count <span class="op">/</span><span class="st"> </span><span class="kw">sum</span>(genes<span class="op">$</span>Count) 
genes</code></pre>
<pre><code>##     SYMBOL GeneLength Count Count_2   Prop_Count
## 1  Gm10568       1634     0      10 0.0000000000
## 2  Gm19860        799     4      14 0.0016214025
## 3  Gm19938       3259     0      10 0.0000000000
## 4   Lypla1       2433   768     778 0.3113092825
## 5      Rp1       9747     0      10 0.0000000000
## 6    Sox17       3130     1      11 0.0004053506
## 7    Tcea1       2847   810     820 0.3283340089
## 8   Mrpl15       4203   431     441 0.1747061208
## 9     Xkr4       3634     1      11 0.0004053506
## 10   Rgs20       2241   452     462 0.1832184840</code></pre>

</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="index.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="session-2.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/lunr.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"google": false,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"all": ["facebook", "google", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": "https://github.com/rstudio/bookdown-demo/edit/master/01-session1.Rmd",
"text": "Edit"
},
"history": {
"link": null,
"text": null
},
"download": ["IntroBaseR.pdf", "IntroBaseR.epub"],
"toc": {
"collapse": "subsection"
}
});
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "true";
    if (src === "" || src === "true") src = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:" && /^https?:/.test(src))
      src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>
</body>

</html>
