<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>Chapter 3 Session 3 | Introduction to Base R</title>
  <meta name="description" content="Introductory tutorial to base R, with a biological focus. In particular it will focus on analysing RNA-sequencing data with Bioconductor." />
  <meta name="generator" content="bookdown 0.10 and GitBook 2.6.7" />

  <meta property="og:title" content="Chapter 3 Session 3 | Introduction to Base R" />
  <meta property="og:type" content="book" />
  
  
  <meta property="og:description" content="Introductory tutorial to base R, with a biological focus. In particular it will focus on analysing RNA-sequencing data with Bioconductor." />
  <meta name="github-repo" content="lucyleeow/BaseR_Intro" />

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Chapter 3 Session 3 | Introduction to Base R" />
  
  <meta name="twitter:description" content="Introductory tutorial to base R, with a biological focus. In particular it will focus on analysing RNA-sequencing data with Bioconductor." />
  

<meta name="author" content="Lucy Liu" />


<meta name="date" content="2019-05-22" />

  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="session-2.html">
<link rel="next" href="session-4.html">
<script src="libs/jquery-2.2.3/jquery.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />







<script src="libs/kePrint-0.0.1/kePrint.js"></script>


<style type="text/css">
div.sourceCode { overflow-x: auto; }
table.sourceCode, tr.sourceCode, td.lineNumbers, td.sourceCode {
  margin: 0; padding: 0; vertical-align: baseline; border: none; }
table.sourceCode { width: 100%; line-height: 100%; }
td.lineNumbers { text-align: right; padding-right: 4px; padding-left: 4px; color: #aaaaaa; border-right: 1px solid #aaaaaa; }
td.sourceCode { padding-left: 5px; }
code > span.kw { color: #007020; font-weight: bold; } /* Keyword */
code > span.dt { color: #902000; } /* DataType */
code > span.dv { color: #40a070; } /* DecVal */
code > span.bn { color: #40a070; } /* BaseN */
code > span.fl { color: #40a070; } /* Float */
code > span.ch { color: #4070a0; } /* Char */
code > span.st { color: #4070a0; } /* String */
code > span.co { color: #60a0b0; font-style: italic; } /* Comment */
code > span.ot { color: #007020; } /* Other */
code > span.al { color: #ff0000; font-weight: bold; } /* Alert */
code > span.fu { color: #06287e; } /* Function */
code > span.er { color: #ff0000; font-weight: bold; } /* Error */
code > span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
code > span.cn { color: #880000; } /* Constant */
code > span.sc { color: #4070a0; } /* SpecialChar */
code > span.vs { color: #4070a0; } /* VerbatimString */
code > span.ss { color: #bb6688; } /* SpecialString */
code > span.im { } /* Import */
code > span.va { color: #19177c; } /* Variable */
code > span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code > span.op { color: #666666; } /* Operator */
code > span.bu { } /* BuiltIn */
code > span.ex { } /* Extension */
code > span.pp { color: #bc7a00; } /* Preprocessor */
code > span.at { color: #7d9029; } /* Attribute */
code > span.do { color: #ba2121; font-style: italic; } /* Documentation */
code > span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code > span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code > span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">Introduction to Base R</a></li>

<li class="divider"></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i>Preface</a><ul>
<li class="chapter" data-level="0.1" data-path="index.html"><a href="index.html#data-files"><i class="fa fa-check"></i><b>0.1</b> Data files</a></li>
</ul></li>
<li class="chapter" data-level="1" data-path="session-1.html"><a href="session-1.html"><i class="fa fa-check"></i><b>1</b> Session 1</a><ul>
<li class="chapter" data-level="1.1" data-path="session-1.html"><a href="session-1.html#r-and-rstudio"><i class="fa fa-check"></i><b>1.1</b> R and RStudio</a></li>
<li class="chapter" data-level="1.2" data-path="session-1.html"><a href="session-1.html#r-scripts"><i class="fa fa-check"></i><b>1.2</b> R scripts</a></li>
<li class="chapter" data-level="1.3" data-path="session-1.html"><a href="session-1.html#help"><i class="fa fa-check"></i><b>1.3</b> Help</a></li>
<li class="chapter" data-level="1.4" data-path="session-1.html"><a href="session-1.html#working-directory"><i class="fa fa-check"></i><b>1.4</b> Working directory</a></li>
<li class="chapter" data-level="1.5" data-path="session-1.html"><a href="session-1.html#maths"><i class="fa fa-check"></i><b>1.5</b> Maths</a></li>
<li class="chapter" data-level="1.6" data-path="session-1.html"><a href="session-1.html#comps"><i class="fa fa-check"></i><b>1.6</b> Comparisons</a></li>
<li class="chapter" data-level="1.7" data-path="session-1.html"><a href="session-1.html#variables"><i class="fa fa-check"></i><b>1.7</b> Variables</a><ul>
<li class="chapter" data-level="1.7.1" data-path="session-1.html"><a href="session-1.html#errors-and-warnings"><i class="fa fa-check"></i><b>1.7.1</b> Errors and warnings</a></li>
</ul></li>
<li class="chapter" data-level="1.8" data-path="session-1.html"><a href="session-1.html#data-structures"><i class="fa fa-check"></i><b>1.8</b> Data structures</a><ul>
<li class="chapter" data-level="1.8.1" data-path="session-1.html"><a href="session-1.html#vector"><i class="fa fa-check"></i><b>1.8.1</b> Vector</a></li>
<li class="chapter" data-level="1.8.2" data-path="session-1.html"><a href="session-1.html#matrix"><i class="fa fa-check"></i><b>1.8.2</b> Matrix</a></li>
<li class="chapter" data-level="1.8.3" data-path="session-1.html"><a href="session-1.html#dataframe"><i class="fa fa-check"></i><b>1.8.3</b> Dataframe</a></li>
</ul></li>
<li class="chapter" data-level="1.9" data-path="session-1.html"><a href="session-1.html#reading-in-data"><i class="fa fa-check"></i><b>1.9</b> Reading in data</a><ul>
<li class="chapter" data-level="1.9.1" data-path="session-1.html"><a href="session-1.html#summary"><i class="fa fa-check"></i><b>1.9.1</b> Summary</a></li>
<li class="chapter" data-level="1.9.2" data-path="session-1.html"><a href="session-1.html#structure"><i class="fa fa-check"></i><b>1.9.2</b> Structure</a></li>
</ul></li>
<li class="chapter" data-level="1.10" data-path="session-1.html"><a href="session-1.html#data-types"><i class="fa fa-check"></i><b>1.10</b> Data types</a><ul>
<li class="chapter" data-level="1.10.1" data-path="session-1.html"><a href="session-1.html#factors"><i class="fa fa-check"></i><b>1.10.1</b> Factors</a></li>
<li class="chapter" data-level="1.10.2" data-path="session-1.html"><a href="session-1.html#type-coercion"><i class="fa fa-check"></i><b>1.10.2</b> Type coercion</a></li>
</ul></li>
<li class="chapter" data-level="1.11" data-path="session-1.html"><a href="session-1.html#vect"><i class="fa fa-check"></i><b>1.11</b> Vectorisation</a></li>
<li class="chapter" data-level="1.12" data-path="session-1.html"><a href="session-1.html#write"><i class="fa fa-check"></i><b>1.12</b> Writing out data</a></li>
<li class="chapter" data-level="1.13" data-path="session-1.html"><a href="session-1.html#homework"><i class="fa fa-check"></i><b>1.13</b> Homework</a></li>
<li class="chapter" data-level="1.14" data-path="session-1.html"><a href="session-1.html#answers"><i class="fa fa-check"></i><b>1.14</b> Answers</a></li>
</ul></li>
<li class="chapter" data-level="2" data-path="session-2.html"><a href="session-2.html"><i class="fa fa-check"></i><b>2</b> Session 2</a><ul>
<li class="chapter" data-level="2.1" data-path="session-2.html"><a href="session-2.html#functions"><i class="fa fa-check"></i><b>2.1</b> Functions</a></li>
<li class="chapter" data-level="2.2" data-path="session-2.html"><a href="session-2.html#miss"><i class="fa fa-check"></i><b>2.2</b> Missing values</a><ul>
<li class="chapter" data-level="2.2.1" data-path="session-2.html"><a href="session-2.html#missing-codes"><i class="fa fa-check"></i><b>2.2.1</b> Missing codes</a></li>
</ul></li>
<li class="chapter" data-level="2.3" data-path="session-2.html"><a href="session-2.html#subsetting"><i class="fa fa-check"></i><b>2.3</b> Subsetting</a><ul>
<li class="chapter" data-level="2.3.1" data-path="session-2.html"><a href="session-2.html#subsetting-with-indicies"><i class="fa fa-check"></i><b>2.3.1</b> Subsetting with indicies</a></li>
<li class="chapter" data-level="2.3.2" data-path="session-2.html"><a href="session-2.html#subsetting-with-logicals"><i class="fa fa-check"></i><b>2.3.2</b> Subsetting with logicals</a></li>
<li class="chapter" data-level="2.3.3" data-path="session-2.html"><a href="session-2.html#in"><i class="fa fa-check"></i><b>2.3.3</b> %in%</a></li>
</ul></li>
<li class="chapter" data-level="2.4" data-path="session-2.html"><a href="session-2.html#merge"><i class="fa fa-check"></i><b>2.4</b> Merge</a></li>
<li class="chapter" data-level="2.5" data-path="session-2.html"><a href="session-2.html#homework-1"><i class="fa fa-check"></i><b>2.5</b> Homework</a></li>
<li class="chapter" data-level="2.6" data-path="session-2.html"><a href="session-2.html#answers-1"><i class="fa fa-check"></i><b>2.6</b> Answers</a></li>
</ul></li>
<li class="chapter" data-level="3" data-path="session-3.html"><a href="session-3.html"><i class="fa fa-check"></i><b>3</b> Session 3</a><ul>
<li class="chapter" data-level="3.1" data-path="session-3.html"><a href="session-3.html#lists"><i class="fa fa-check"></i><b>3.1</b> Lists</a><ul>
<li class="chapter" data-level="3.1.1" data-path="session-3.html"><a href="session-3.html#named-list"><i class="fa fa-check"></i><b>3.1.1</b> Named list</a></li>
</ul></li>
<li class="chapter" data-level="3.2" data-path="session-3.html"><a href="session-3.html#classes"><i class="fa fa-check"></i><b>3.2</b> Classes</a></li>
<li class="chapter" data-level="3.3" data-path="session-3.html"><a href="session-3.html#packages"><i class="fa fa-check"></i><b>3.3</b> Packages</a></li>
<li class="chapter" data-level="3.4" data-path="session-3.html"><a href="session-3.html#dgelist"><i class="fa fa-check"></i><b>3.4</b> DGEList</a></li>
<li class="chapter" data-level="3.5" data-path="session-3.html"><a href="session-3.html#homework-2"><i class="fa fa-check"></i><b>3.5</b> Homework</a></li>
<li class="chapter" data-level="3.6" data-path="session-3.html"><a href="session-3.html#answers-2"><i class="fa fa-check"></i><b>3.6</b> Answers</a></li>
</ul></li>
<li class="chapter" data-level="4" data-path="session-4.html"><a href="session-4.html"><i class="fa fa-check"></i><b>4</b> Session 4</a><ul>
<li class="chapter" data-level="4.1" data-path="session-4.html"><a href="session-4.html#filtering"><i class="fa fa-check"></i><b>4.1</b> Filtering</a></li>
<li class="chapter" data-level="4.2" data-path="session-4.html"><a href="session-4.html#normalisation"><i class="fa fa-check"></i><b>4.2</b> Normalisation</a></li>
<li class="chapter" data-level="4.3" data-path="session-4.html"><a href="session-4.html#mds"><i class="fa fa-check"></i><b>4.3</b> MDS</a><ul>
<li class="chapter" data-level="4.3.1" data-path="session-4.html"><a href="session-4.html#plotting"><i class="fa fa-check"></i><b>4.3.1</b> Plotting</a></li>
</ul></li>
<li class="chapter" data-level="4.4" data-path="session-4.html"><a href="session-4.html#saving-plots"><i class="fa fa-check"></i><b>4.4</b> Saving plots</a></li>
<li class="chapter" data-level="4.5" data-path="session-4.html"><a href="session-4.html#linear-modelling"><i class="fa fa-check"></i><b>4.5</b> Linear modelling</a><ul>
<li class="chapter" data-level="4.5.1" data-path="session-4.html"><a href="session-4.html#design-matrix"><i class="fa fa-check"></i><b>4.5.1</b> Design matrix</a></li>
<li class="chapter" data-level="4.5.2" data-path="session-4.html"><a href="session-4.html#contrasts"><i class="fa fa-check"></i><b>4.5.2</b> Contrasts</a></li>
<li class="chapter" data-level="4.5.3" data-path="session-4.html"><a href="session-4.html#linear-model-fitting"><i class="fa fa-check"></i><b>4.5.3</b> Linear model fitting</a></li>
</ul></li>
<li class="chapter" data-level="4.6" data-path="session-4.html"><a href="session-4.html#statistical-testing"><i class="fa fa-check"></i><b>4.6</b> Statistical testing</a></li>
<li class="chapter" data-level="4.7" data-path="session-4.html"><a href="session-4.html#answers-3"><i class="fa fa-check"></i><b>4.7</b> Answers</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="references.html"><a href="references.html"><i class="fa fa-check"></i>References</a></li>
<li class="divider"></li>
<li><a href="https://github.com/rstudio/bookdown" target="blank">Published with bookdown</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Introduction to Base R</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="session-3" class="section level1">
<h1><span class="header-section-number">Chapter 3</span> Session 3</h1>
<p>In this session we will learn:</p>
<ul>
<li>lists, another data structure</li>
<li>Classes</li>
<li>the DGEList object from the package limma</li>
</ul>
<div id="lists" class="section level2">
<h2><span class="header-section-number">3.1</span> Lists</h2>
<p>Recall that we have previously learnt about the data structures: vectors, matrices and dataframes. Another important data structure is the list. Like a vector, it is 1 dimensional i.e. one row of data. Unlike vectors, you can put several data types in a list. Here, our list includes data of the integer, a character and a double types:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">list</span>(<span class="dv">1</span>, <span class="st">&quot;a&quot;</span>, <span class="fl">1.5</span>)</code></pre></div>
<pre><code>## [[1]]
## [1] 1
## 
## [[2]]
## [1] &quot;a&quot;
## 
## [[3]]
## [1] 1.5</code></pre>
<p>Not only can you put different data types into a list, you can also put a WHOLE data structure into one element of a list. In the list below, the first element is a vector that contains 3 numbers, the second element is a character and the third element is a dataframe that has two columns.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">list</span>(<span class="kw">c</span>(<span class="dv">1</span>,<span class="dv">2</span>,<span class="dv">3</span>), 
     <span class="kw">c</span>(<span class="st">&quot;words&quot;</span>, <span class="st">&quot;letters&quot;</span>), 
     <span class="kw">data.frame</span>(<span class="dt">column1 =</span> <span class="kw">c</span>(<span class="dv">1</span>,<span class="dv">2</span>,<span class="dv">3</span>), <span class="dt">column2 =</span> <span class="kw">c</span>(<span class="st">&quot;a&quot;</span>,<span class="st">&quot;b&quot;</span>,<span class="st">&quot;c&quot;</span>))
     )</code></pre></div>
<pre><code>## [[1]]
## [1] 1 2 3
## 
## [[2]]
## [1] &quot;words&quot;   &quot;letters&quot;
## 
## [[3]]
##   column1 column2
## 1       1       a
## 2       2       b
## 3       3       c</code></pre>
<p>The output can often help you understand how the list is structured. The double brackets (e.g. <code>[[1]]</code>) signifies an element of the list and which index it is at. Here there are three elements in our list so the numbers in the double square brackets go from 1 to 3. Underneath the <code>[[1]]</code> and <code>[[2]]</code>, there is a <code>[1]</code> - this indicates that the first and second elements both contain a vector. Underneath <code>[[3]]</code> you see the standard output for a dataframe, which we have seen before.</p>
<p>You even include a list within a list (within a list, within a list….I call this ‘list - ception’). This is where it starts to get confusing.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">list</span>(<span class="dv">1</span>, <span class="kw">list</span>(<span class="dv">1</span>,<span class="st">&quot;a&quot;</span>))</code></pre></div>
<pre><code>## [[1]]
## [1] 1
## 
## [[2]]
## [[2]][[1]]
## [1] 1
## 
## [[2]][[2]]
## [1] &quot;a&quot;</code></pre>
<ul>
<li>The first element, indicated by <code>[[1]]</code>, is a vector, indicated by the <code>[1]</code> underneath.</li>
<li>The second element, indicated by the first <code>[[2]]</code> contains a list:
<ul>
<li><code>[[2]][[1]]</code> - tells you that the second element is a list, of which the first element of the inner list is the number 1.</li>
<li><code>[[2]][[2]]</code> - tells you that the second element is a list, of which the second element of the inner list is “a”.</li>
</ul></li>
</ul>
<p>Many bioconductor packages, including limma, use lists so it is an important data structure to understand.</p>
<p><strong>Challenge 3.1</strong></p>
<p>Below is the output from a list.</p>
<ol style="list-style-type: decimal">
<li><p>How many element of the list are there?</p></li>
<li><p>Look carefully at each element and answer the following questions about EVERY element of the list:</p></li>
</ol>
<ul>
<li>How many elements does the element of the list contain?</li>
<li>If there are several elements within this element, what does each element contain?</li>
</ul>
<pre><code>## [[1]]
## [[1]][[1]]
## [1] 1 2
## 
## [[1]][[2]]
## [1] &quot;b&quot;
## 
## 
## [[2]]
## [1] &quot;a&quot; &quot;b&quot;
## 
## [[3]]
## [[3]][[1]]
## [[3]][[1]][[1]]
## [1] &quot;a&quot;
## 
## [[3]][[1]][[2]]
## [1] 1
## 
## 
## [[3]][[2]]
## [1] &quot;b&quot;
## 
## 
## [[4]]
## [1] &quot;end&quot;</code></pre>
<p>We can access elements of a list using square brackets. You may have noticed that there are often double square brackets <code>[[ ]]</code>. There is a subtle but important difference between single square brackets and double square brackets when subsetting lists.</p>
<p>Let’s make a list:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">my_list &lt;-<span class="st"> </span><span class="kw">list</span>(
  <span class="kw">c</span>(<span class="st">&quot;a&quot;</span>,<span class="st">&quot;b&quot;</span>),
  <span class="kw">c</span>(<span class="dv">1</span>,<span class="dv">2</span>,<span class="dv">3</span>),
  <span class="kw">c</span>(<span class="dv">4</span>,<span class="dv">5</span>,<span class="dv">6</span>)
  )</code></pre></div>
<p>Take a look at the output of obtaining the first element with single square brackets and double square brackets:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">my_list[<span class="dv">1</span>]</code></pre></div>
<pre><code>## [[1]]
## [1] &quot;a&quot; &quot;b&quot;</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">my_list[[<span class="dv">1</span>]]</code></pre></div>
<pre><code>## [1] &quot;a&quot; &quot;b&quot;</code></pre>
<p>The difference is minor; with <code>my_list[1]</code> there is an extra <code>[[1]]</code> at the top, with <code>my_list[[1]]</code> there is no <code>[[1]]</code> at the top.</p>
<p>The <code>str()</code> function gives us some more information:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">str</span>(my_list[<span class="dv">1</span>])</code></pre></div>
<pre><code>## List of 1
##  $ : chr [1:2] &quot;a&quot; &quot;b&quot;</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">str</span>(my_list[[<span class="dv">1</span>]])</code></pre></div>
<pre><code>##  chr [1:2] &quot;a&quot; &quot;b&quot;</code></pre>
<p>This tells us that when you use single brackets, the output is a list containing one element, which is a vector. When you use double brackets, the output is just a vector, NOT contained within a list.</p>
<p>Let’s use an analogy to gain a better understanding of this concept. Below is a picture of a cargo train, which contains a box in each segment. This represents a list containing 3 elements, with each element being the box.</p>
<div class="figure" style="text-align: center"><span id="fig:train"></span>
<img src="Images/train.jpg" alt="Cargo train representation of a list." width="700px" />
<p class="caption">
Figure 3.1: Cargo train representation of a list.
</p>
</div>
<p>Using a single bracket returns you the train segment with the box inside.</p>
<div class="figure" style="text-align: center"><span id="fig:trainS"></span>
<img src="Images/train_double.jpeg" alt="Single brackets with our cargo train list." width="300px" height="350px" />
<p class="caption">
Figure 3.2: Single brackets with our cargo train list.
</p>
</div>
<p>Using double brackets returns you just the box inside.</p>
<div class="figure" style="text-align: center"><span id="fig:trainD"></span>
<img src="Images/box.png" alt="Double brackets with our cargo train list." width="300px" />
<p class="caption">
Figure 3.3: Double brackets with our cargo train list.
</p>
</div>
<p>Back to our list:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">my_list</code></pre></div>
<pre><code>## [[1]]
## [1] &quot;a&quot; &quot;b&quot;
## 
## [[2]]
## [1] 1 2 3
## 
## [[3]]
## [1] 4 5 6</code></pre>
<p>To obtain the first element of the vector contained within the first element of the list (the <code>&quot;a&quot;</code>) you can use:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">my_list[[<span class="dv">1</span>]][<span class="dv">1</span>]</code></pre></div>
<pre><code>## [1] &quot;a&quot;</code></pre>
<p>The <code>[[1]]</code> gives you just the vector contained within the first element of the list:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">my_list[[<span class="dv">1</span>]]</code></pre></div>
<pre><code>## [1] &quot;a&quot; &quot;b&quot;</code></pre>
<p>The second square bracket <code>[1]</code> then gives you the first element of the vector:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">my_list[[<span class="dv">1</span>]][<span class="dv">1</span>]</code></pre></div>
<pre><code>## [1] &quot;a&quot;</code></pre>
<p><strong>Challenge 3.2</strong></p>
<p>First create a new list:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">my_list2 &lt;-<span class="st"> </span><span class="kw">list</span>(
  <span class="kw">c</span>(<span class="st">&quot;a&quot;</span>,<span class="st">&quot;b&quot;</span>),
  <span class="kw">c</span>(<span class="dv">1</span>,<span class="dv">2</span>,<span class="dv">3</span>),
  <span class="kw">data.frame</span>(<span class="dt">Name =</span> <span class="kw">c</span>(<span class="st">&quot;Sally&quot;</span>, <span class="st">&quot;Andy&quot;</span>), <span class="dt">Age =</span> <span class="kw">c</span>(<span class="dv">40</span>,<span class="dv">32</span>))
  )
my_list2</code></pre></div>
<pre><code>## [[1]]
## [1] &quot;a&quot; &quot;b&quot;
## 
## [[2]]
## [1] 1 2 3
## 
## [[3]]
##    Name Age
## 1 Sally  40
## 2  Andy  32</code></pre>
<p>Subset this list to obtain a data structure that gives the following outputs:</p>
<ol style="list-style-type: decimal">
<li></li>
</ol>
<pre><code>## [1] &quot;a&quot; &quot;b&quot;</code></pre>
<ol start="2" style="list-style-type: decimal">
<li></li>
</ol>
<pre><code>## [1] &quot;b&quot;</code></pre>
<ol start="3" style="list-style-type: decimal">
<li></li>
</ol>
<pre><code>## [[1]]
## [1] 1 2 3</code></pre>
<ol start="4" style="list-style-type: decimal">
<li></li>
</ol>
<pre><code>## [1] 40</code></pre>
<ol start="5" style="list-style-type: decimal">
<li></li>
</ol>
<pre><code>##    Name Age
## 1 Sally  40</code></pre>
<ol start="6" style="list-style-type: decimal">
<li></li>
</ol>
<pre><code>## [[1]]
##    Name Age
## 1 Sally  40
## 2  Andy  32</code></pre>
<div id="named-list" class="section level3">
<h3><span class="header-section-number">3.1.1</span> Named list</h3>
<p>You can also have names for each element of your list:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">named_list &lt;-<span class="st"> </span><span class="kw">list</span>(
  <span class="dt">name =</span> <span class="kw">c</span>(<span class="st">&quot;Andy&quot;</span>, <span class="st">&quot;Sally&quot;</span>),
  <span class="dt">age =</span> <span class="kw">c</span>(<span class="dv">32</span>,<span class="dv">40</span>)
)
named_list</code></pre></div>
<pre><code>## $name
## [1] &quot;Andy&quot;  &quot;Sally&quot;
## 
## $age
## [1] 32 40</code></pre>
<p>If your list is named, you can use <code>$</code> to access each element of your list:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">named_list<span class="op">$</span>name</code></pre></div>
<pre><code>## [1] &quot;Andy&quot;  &quot;Sally&quot;</code></pre>
<p>Note that the output is just the vector, NOT a vector within a list. Thus <code>named_list[[1]]</code> does the same thing as <code>named_list$name</code>.</p>
<p>Recall that we access columns in dataframes with the <code>$</code> symbol as well. This means that if you have a dataframe within a list, you can obtain a column within the dataframe that is within the list using two <code>$</code>’s:</p>
<p>Let’s start with a list that contains a dataframe as its first element</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">named_list2 &lt;-<span class="st"> </span><span class="kw">list</span>(
 <span class="dt">details =</span> <span class="kw">data.frame</span>(<span class="dt">name =</span> <span class="kw">c</span>(<span class="st">&quot;Andy&quot;</span>, <span class="st">&quot;Sally&quot;</span>),
                      <span class="dt">age =</span> <span class="kw">c</span>(<span class="dv">32</span>,<span class="dv">40</span>))
)
named_list2</code></pre></div>
<pre><code>## $details
##    name age
## 1  Andy  32
## 2 Sally  40</code></pre>
<p>We can get the age column using:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">named_list2<span class="op">$</span>details<span class="op">$</span>age</code></pre></div>
<pre><code>## [1] 32 40</code></pre>
<p><em>Extra</em>:The reason you can also access columns in a dataframe with <code>$</code> is because ‘under the hood’ a dataframe is actually a list. It is a list with the constraint that each element is a vector of the same length. Each element in the list is thus a column in the ‘dataframe’.</p>
<p><strong>Challenge 3.3</strong></p>
<p>Take a look at the named list below:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">named_list3 &lt;-<span class="st"> </span><span class="kw">list</span>(
  <span class="dt">cats =</span> <span class="kw">data.frame</span>(<span class="dt">name =</span> <span class="kw">c</span>(<span class="st">&quot;Garfield&quot;</span>, <span class="st">&quot;Hello Kitty&quot;</span>),
                    <span class="dt">age =</span> <span class="kw">c</span>(<span class="dv">3</span>,<span class="dv">10</span>), <span class="dt">stringsAsFactors =</span> <span class="ot">FALSE</span>),
  <span class="dt">dogs =</span> <span class="kw">data.frame</span>(<span class="dt">name =</span> <span class="kw">c</span>(<span class="st">&quot;Spot&quot;</span>, <span class="st">&quot;Snoopy&quot;</span>),
                    <span class="dt">age =</span> <span class="kw">c</span>(<span class="dv">5</span>,<span class="dv">14</span>), <span class="dt">stringsAsFactors =</span> <span class="ot">FALSE</span>)
)
named_list3</code></pre></div>
<pre><code>## $cats
##          name age
## 1    Garfield   3
## 2 Hello Kitty  10
## 
## $dogs
##     name age
## 1   Spot   5
## 2 Snoopy  14</code></pre>
<p>Using just <code>$</code>, obtain:</p>
<ol style="list-style-type: decimal">
<li>the vector of cat names</li>
<li>the dog dataframe</li>
<li>the vector of dog ages</li>
</ol>
</div>
</div>
<div id="classes" class="section level2">
<h2><span class="header-section-number">3.2</span> Classes</h2>
<p>Everything in R is an ‘object’ - every function that we have used and every data structure we have created in R. Each object falls under a ‘class’. A class defines a type of object and what properties it has. For example, every list created with <code>list()</code> is an object that falls under the class ‘list’. The ‘list’ class describes how ‘list’ objects behave.</p>
<p>To find out what class an object falls under, use the function <code>class()</code>:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">class</span>(<span class="kw">list</span>(<span class="dv">1</span>, <span class="dv">2</span>, <span class="st">&quot;a&quot;</span>))</code></pre></div>
<pre><code>## [1] &quot;list&quot;</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">class</span>(<span class="kw">list</span>(<span class="st">&quot;a&quot;</span>, <span class="st">&quot;b&quot;</span>, <span class="dv">4</span>))</code></pre></div>
<pre><code>## [1] &quot;list&quot;</code></pre>
<p>Bioconductor packages often use special classes. For example, the limma package uses the <code>DGEList</code> class. This is a class that is specifically designed for storing read count data from RNA sequencing. It is a special ‘list’ that must contain two components:</p>
<ul>
<li><code>counts</code> - which must be a numeric matrix, that stores counts. Each row must be a gene and each column must be a sample.</li>
<li><code>samples</code> - which must be a dataframe, that contains information about each sample. Each row must be a sample and must contain information about the group (e.g. treatment group) the sample belongs to, the library size of that sample and the normalisation factor for that sample.</li>
</ul>
<p>There are also a number of optional components of the <code>DGEList</code> class, such as a dataframe containing gene annotation information.</p>
</div>
<div id="packages" class="section level2">
<h2><span class="header-section-number">3.3</span> Packages</h2>
<p>Last session we installed the packages limma and edgeR. This downloads the files for each package and saves them to your computer. You generally only need to do this once.</p>
<p>To use a package you must ‘load’ them EACH time you start a new R session. You do this with the <code>library()</code> function. Let’s load both limma and edgeR:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(edgeR)</code></pre></div>
<pre><code>## Loading required package: limma</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(limma)</code></pre></div>
</div>
<div id="dgelist" class="section level2">
<h2><span class="header-section-number">3.4</span> DGEList</h2>
<p>The RNA sequencing analysis you will be guided through is a simplified version of that performed in the article from Law <em>et al.</em> <span class="citation">(Law et al. <a href="#ref-law2016rna">2016</a>)</span>. The RNA sequencing data we will use is from Sheridan <em>et al.</em> <span class="citation">(Sheridan et al. <a href="#ref-sheridan2015pooled">2015</a>)</span>. It consists of samples from 3 cell populations; basal, luminal progenitor (LP) and mature luminal (ML), which has been sorted from the mammary glands of female virgin mice. The reads have been aligned to the mouse reference genome (mm10) and reads summarised at the gene-level (using mm10 RefSeq-based annotation) to obtain gene counts. Gene level summarisation involves counting the number of reads mapped to each gene - the resulting ‘count of reads’ is often referred to simply as ‘counts’.</p>
<p>We are going to start our RNA-seq analysis with gene counts. The data files for this session should have been emailed to you (though you can also obtain them from <a href="https://github.com/lucyleeow/BaseR_Intro">Github</a>). Please uncompress (extract) the files and put them in your working directory.</p>
<p>Each data file details the number of reads mapped to a gene for 1 sample. There is 1 data file for each sample. Within each data file, there are 3 columns - ‘EntrezID’, ‘GeneLength’ and ‘Count.’</p>
<p>We will be looking at 9 samples (and using 9 data files) in total. Their details are shown below:</p>
<table>
<caption><span id="tab:samps">Table 3.1: </span>Summary of samples.</caption>
<thead>
<tr class="header">
<th align="left">File name</th>
<th align="left">Sample name</th>
<th align="left">Phenotype group</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="left">GSM1545535_10_6_5_11.txt</td>
<td align="left">10_6_5_11</td>
<td align="left">LP</td>
</tr>
<tr class="even">
<td align="left">GSM1545536_9_6_5_11.txt</td>
<td align="left">9_6_5_11</td>
<td align="left">ML</td>
</tr>
<tr class="odd">
<td align="left">GSM1545538_purep53.txt</td>
<td align="left">purep53</td>
<td align="left">Basal</td>
</tr>
<tr class="even">
<td align="left">GSM1545539_JMS8-2.txt</td>
<td align="left">JMS8-2</td>
<td align="left">Basal</td>
</tr>
<tr class="odd">
<td align="left">GSM1545540_JMS8-3.txt</td>
<td align="left">JMS8-3</td>
<td align="left">ML</td>
</tr>
<tr class="even">
<td align="left">GSM1545541_JMS8-4.txt</td>
<td align="left">JMS8-4</td>
<td align="left">LP</td>
</tr>
<tr class="odd">
<td align="left">GSM1545542_JMS8-5.txt</td>
<td align="left">JMS8-5</td>
<td align="left">Basal</td>
</tr>
<tr class="even">
<td align="left">GSM1545544_JMS9-P7c.txt</td>
<td align="left">JMS9-P7c</td>
<td align="left">ML</td>
</tr>
<tr class="odd">
<td align="left">GSM1545545_JMS9-P8c.txt</td>
<td align="left">JMS9-P8c</td>
<td align="left">LP</td>
</tr>
</tbody>
</table>
<p>To create a <code>DGEList</code> class object (or simply ‘<code>DGEList</code> object’), we will use the <code>readDGE()</code> function. There are three important arguments to this function:</p>
<ul>
<li><code>files</code> - a vector of data file names</li>
<li><code>path</code> - the path to the directory that contains your data files. If the data files are in your working directory, don’t worry about this argument. If the data files are somewhere else, like a folder called ‘data’, in your working directory you must give the path to that folder</li>
<li><code>colums</code> - the columns of the input files which have the gene names and counts respectively (as the column indices)</li>
</ul>
<p>First, we will create a vector of the file names. You can simply copy and paste this code into your R script.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">files &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;GSM1545535_10_6_5_11.txt&quot;</span>, <span class="st">&quot;GSM1545536_9_6_5_11.txt&quot;</span>, 
   <span class="st">&quot;GSM1545538_purep53.txt&quot;</span>, <span class="st">&quot;GSM1545539_JMS8-2.txt&quot;</span>, 
   <span class="st">&quot;GSM1545540_JMS8-3.txt&quot;</span>, <span class="st">&quot;GSM1545541_JMS8-4.txt&quot;</span>, 
   <span class="st">&quot;GSM1545542_JMS8-5.txt&quot;</span>, <span class="st">&quot;GSM1545544_JMS9-P7c.txt&quot;</span>, 
   <span class="st">&quot;GSM1545545_JMS9-P8c.txt&quot;</span>)</code></pre></div>
<p>Next, we will create our <code>DGEList</code> object. I have put my data files in a folder called “data” (within my working directory). Thus, I must specify <code>path = &quot;data&quot;</code>. Depending on where you have put your data files, you may need a different input to <code>path</code> or not have to include the <code>path</code> argument.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">x &lt;-<span class="st"> </span><span class="kw">readDGE</span>(files, <span class="dt">path =</span> <span class="st">&quot;data&quot;</span>, <span class="dt">columns =</span> <span class="kw">c</span>(<span class="dv">1</span>,<span class="dv">3</span>))</code></pre></div>
<p>The object we created has 2 elements, one named <code>samples</code> and one named <code>counts</code>. Let’s take a look at each:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">head</span>(x<span class="op">$</span>counts)</code></pre></div>
<pre><code>##            Samples
## Tags        GSM1545535_10_6_5_11 GSM1545536_9_6_5_11 GSM1545538_purep53
##   497097                       1                   2                342
##   100503874                    0                   0                  5
##   100038431                    0                   0                  0
##   19888                        0                   1                  0
##   20671                        1                   1                 76
##   27395                      431                 771               1368
##            Samples
## Tags        GSM1545539_JMS8-2 GSM1545540_JMS8-3 GSM1545541_JMS8-4
##   497097                  526                 3                 3
##   100503874                 6                 0                 0
##   100038431                 0                 0                 0
##   19888                     0                17                 2
##   20671                    40                33                14
##   27395                  1268              1564               769
##            Samples
## Tags        GSM1545542_JMS8-5 GSM1545544_JMS9-P7c GSM1545545_JMS9-P8c
##   497097                  535                   2                   0
##   100503874                 5                   0                   0
##   100038431                 1                   0                   0
##   19888                     0                   1                   0
##   20671                    98                  18                   8
##   27395                   818                 468                 342</code></pre>
<p>Recall that <code>DGEList</code> specifies that the <code>counts</code> element of the list must be a numeric matrix, that stores the counts. Each row must be a gene and each column must be a sample. Note that the column names are the file names and the rownames are the EntrezID’s.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">head</span>(x<span class="op">$</span>samples)</code></pre></div>
<pre><code>##                                         files group lib.size norm.factors
## GSM1545535_10_6_5_11 GSM1545535_10_6_5_11.txt     1 32863052            1
## GSM1545536_9_6_5_11   GSM1545536_9_6_5_11.txt     1 35335491            1
## GSM1545538_purep53     GSM1545538_purep53.txt     1 57160817            1
## GSM1545539_JMS8-2       GSM1545539_JMS8-2.txt     1 51368625            1
## GSM1545540_JMS8-3       GSM1545540_JMS8-3.txt     1 75795034            1
## GSM1545541_JMS8-4       GSM1545541_JMS8-4.txt     1 60517657            1</code></pre>
<p>This is a dataframe where each row is 1 sample, and details of each sample is given in the 4 columns.</p>
<ul>
<li><code>files</code> gives the file names.</li>
<li><code>group</code> details the phenotype group the sample belongs to. As we have not specified this yet, the default <code>1</code> is given for each sample.</li>
<li><code>lib.size</code> gives the library size. This is the total sum of all counts for that sample.</li>
<li><code>norm.factors</code> this gives the normalisation factor for each sample. As we have not calculated any normalisation factors, this is 1 for each sample.</li>
</ul>
<p>This dataframe also has rownames, which are currently the file names of each sample. Let’s change them to be the sample names (see Table <a href="session-3.html#tab:samps">3.1</a>).</p>
<p>We can do this by first creating a vector of sample names. You can simply copy and paste the code below into your R script.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">samplenames &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;10_6_5_11&quot;</span>, <span class="st">&quot;9_6_5_11&quot;</span>, <span class="st">&quot;purep53&quot;</span>, <span class="st">&quot;JMS8-2&quot;</span>, <span class="st">&quot;JMS8-3&quot;</span>,
                 <span class="st">&quot;JMS8-4&quot;</span>, <span class="st">&quot;JMS8-5&quot;</span>, <span class="st">&quot;JMS9-P7c&quot;</span>, <span class="st">&quot;JMS9-P8c&quot;</span>)</code></pre></div>
<p>We can change the rownames in the <code>samples</code> dataframe using the <code>rownames()</code> function. This function will give you the row names of a data structure (a dataframe in this instance):</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">rownames</span>(x<span class="op">$</span>samples) </code></pre></div>
<pre><code>## [1] &quot;GSM1545535_10_6_5_11&quot; &quot;GSM1545536_9_6_5_11&quot;  &quot;GSM1545538_purep53&quot;  
## [4] &quot;GSM1545539_JMS8-2&quot;    &quot;GSM1545540_JMS8-3&quot;    &quot;GSM1545541_JMS8-4&quot;   
## [7] &quot;GSM1545542_JMS8-5&quot;    &quot;GSM1545544_JMS9-P7c&quot;  &quot;GSM1545545_JMS9-P8c&quot;</code></pre>
<p>We can see that the row names are the file names as we saw above. To replace these file names with the sample names we can run:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">rownames</span>(x<span class="op">$</span>samples) &lt;-<span class="st"> </span>samplenames</code></pre></div>
<p>We have seen this type of notation before in session 1 (section <a href="session-1.html#vect">1.11</a>). This code assigns the <code>samplenames</code> vector as the row names of the <code>samples</code> dataframe.</p>
<p><strong>Challenge 3.4</strong></p>
<ol style="list-style-type: decimal">
<li>Use the <code>colnames()</code> function to replace the column names of the count matrix from being file names to sample names.</li>
<li>The code below creates a factor vector (called <code>group</code>) that specifies the phenotype group each sample belongs to. It is ordered such that the first element in the vector corresponds to the first row of the <code>samples</code> dataframe.</li>
</ol>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">group &lt;-<span class="st"> </span><span class="kw">as.factor</span>(<span class="kw">c</span>(<span class="st">&quot;LP&quot;</span>, <span class="st">&quot;ML&quot;</span>, <span class="st">&quot;Basal&quot;</span>, <span class="st">&quot;Basal&quot;</span>, <span class="st">&quot;ML&quot;</span>, <span class="st">&quot;LP&quot;</span>, <span class="st">&quot;Basal&quot;</span>, <span class="st">&quot;ML&quot;</span>, <span class="st">&quot;LP&quot;</span>)) </code></pre></div>
<p>Replace the <code>group</code> column of the <code>samples</code> dataframe with the <code>group</code> factor vector.</p>
<p><em>Hint</em>: you will need to use the list subsetting notation we learnt at the start of this session.</p>
<ol start="3" style="list-style-type: decimal">
<li>Additional gene annotation information about the genes from our RNA-seq data can be found in the file “Ses3_geneAnnot.tsv”. Read this file in and add it as another element named <code>genes</code> in <code>x</code> (our <code>DGEList</code> object).</li>
</ol>
<p><em>Hint</em>: the command for adding an element to a list is similar to the command for adding a column to a dataframe. Take a look at Section <a href="session-1.html#vect">1.11</a> to review how to do the latter.</p>
</div>
<div id="homework-2" class="section level2">
<h2><span class="header-section-number">3.5</span> Homework</h2>
<ol style="list-style-type: decimal">
<li>Subset the first column of the <code>count</code> matrix from <code>x</code> (our <code>DGEList</code> object) and ‘save’ (assign) it to a variable called <code>sample1</code>.</li>
</ol>
<ul>
<li>Find out how many genes had a count of ‘0’.</li>
<li>Find the total sum of all counts for that sample. Compare this number with the corresponding (first) number in the <code>lib.size</code> column from the <code>samples</code> dataframe. Is it the same?</li>
</ul>
<ol start="2" style="list-style-type: decimal">
<li><p>Calculate the total sum of the library sizes, using the <code>lib.size</code> column from the <code>samples</code> dataframe from <code>x</code> (our <code>DGEList</code> object).</p></li>
<li><p>Using the <code>genes</code> dataframe from <code>x</code> (our <code>DGEList</code> object), find out how many genes are from chromosome 5.</p></li>
</ol>
<p><em>Hint</em>: you will need to remove rows where the <code>TXCHROM</code> column is <code>NA</code>. Revisit section <a href="session-2.html#miss">2.2</a> to review missing values.</p>
</div>
<div id="answers-2" class="section level2">
<h2><span class="header-section-number">3.6</span> Answers</h2>
<p><strong>Challenge 3.1</strong></p>
<ol style="list-style-type: decimal">
<li>There are 4 elements of this list.</li>
<li>For each element -</li>
</ol>
<ul>
<li>Within the first element, there are 2 elements. The first is a vector containing 2 numbers and the second is a vector containing one character type (note there are no ‘scalars’ in R, thus <code>&quot;a&quot;</code> is a vector with 1 element).</li>
<li>The second element contains 1 element. It is a vector containing two character types.</li>
<li>The third element contains 2 elements. Within the first element is another list. Within this list there are 2 elements, both being character types. The second element of this nested list is a vector containing one character type.</li>
<li>The fourth element is a vector containing one character type.</li>
</ul>
<p><strong>Challenge 3.2</strong></p>
<ol style="list-style-type: decimal">
<li></li>
</ol>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">my_list2[[<span class="dv">1</span>]]</code></pre></div>
<pre><code>## [1] &quot;a&quot; &quot;b&quot;</code></pre>
<ol start="2" style="list-style-type: decimal">
<li></li>
</ol>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">my_list2[[<span class="dv">1</span>]][<span class="dv">2</span>]</code></pre></div>
<pre><code>## [1] &quot;b&quot;</code></pre>
<ol start="3" style="list-style-type: decimal">
<li></li>
</ol>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">my_list2[<span class="dv">2</span>]</code></pre></div>
<pre><code>## [[1]]
## [1] 1 2 3</code></pre>
<ol start="4" style="list-style-type: decimal">
<li></li>
</ol>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">my_list2[[<span class="dv">3</span>]][<span class="dv">1</span>,<span class="dv">2</span>]</code></pre></div>
<pre><code>## [1] 40</code></pre>
<ol start="5" style="list-style-type: decimal">
<li></li>
</ol>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">my_list2[[<span class="dv">3</span>]][<span class="dv">1</span>,]</code></pre></div>
<pre><code>##    Name Age
## 1 Sally  40</code></pre>
<ol start="6" style="list-style-type: decimal">
<li></li>
</ol>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">my_list2[<span class="dv">3</span>]</code></pre></div>
<pre><code>## [[1]]
##    Name Age
## 1 Sally  40
## 2  Andy  32</code></pre>
<p><strong>Challenge 3.3</strong></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># 1. the vector of cat names</span>
named_list3<span class="op">$</span>cats<span class="op">$</span>name</code></pre></div>
<pre><code>## [1] &quot;Garfield&quot;    &quot;Hello Kitty&quot;</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># 2. the dog dataframe</span>
named_list3<span class="op">$</span>dogs</code></pre></div>
<pre><code>##     name age
## 1   Spot   5
## 2 Snoopy  14</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># 3. the vector of dog ages</span>
named_list3<span class="op">$</span>dogs<span class="op">$</span>age</code></pre></div>
<pre><code>## [1]  5 14</code></pre>
<p><strong>Challenge 3.4</strong></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># 1</span>
<span class="kw">colnames</span>(x<span class="op">$</span>counts) &lt;-<span class="st"> </span>samplenames
<span class="co"># 2</span>
x<span class="op">$</span>samples<span class="op">$</span>group &lt;-<span class="st"> </span>group
<span class="co"># 3</span>
geneAnnot &lt;-<span class="st"> </span><span class="kw">read.delim</span>(<span class="st">&quot;data/Ses3_geneAnnot.tsv&quot;</span>, <span class="dt">stringsAsFactors =</span> <span class="ot">FALSE</span>)
x<span class="op">$</span>genes &lt;-<span class="st"> </span>geneAnnot</code></pre></div>

</div>
</div>
<h3>References</h3>
<div id="refs" class="references">
<div id="ref-law2016rna">
<p>Law, Charity W, Monther Alhamdoosh, Shian Su, Gordon K Smyth, and Matthew E Ritchie. 2016. “RNA-Seq Analysis Is Easy as 1-2-3 with Limma, Glimma and edgeR.” <em>F1000Research</em> 5. Faculty of 1000 Ltd.</p>
</div>
<div id="ref-sheridan2015pooled">
<p>Sheridan, Julie M, Matthew E Ritchie, Sarah A Best, Kun Jiang, Tamara J Beck, François Vaillant, Kevin Liu, et al. 2015. “A Pooled shRNA Screen for Regulators of Primary Mammary Stem and Progenitor Cells Identifies Roles for Asap1 and Prox1.” <em>BMC Cancer</em> 15 (1). BioMed Central: 221.</p>
</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="session-2.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="session-4.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/lunr.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"google": false,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"all": ["facebook", "google", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": "https://github.com/rstudio/bookdown-demo/edit/master/03-session3.Rmd",
"text": "Edit"
},
"history": {
"link": null,
"text": null
},
"download": ["IntroBaseR.pdf", "IntroBaseR.epub"],
"toc": {
"collapse": "subsection"
}
});
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "";
    if (src === "" || src === "true") src = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:" && /^https?:/.test(src))
      src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>
</body>

</html>
